{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "19437731",
   "metadata": {},
   "outputs": [],
   "source": [
    "from eva_cttv_pipeline.clinvar_xml_utils import *\n",
    "\n",
    "import os\n",
    "import sys\n",
    "from urllib import request\n",
    "import xml.etree.ElementTree as ElementTree\n",
    "\n",
    "import hgvs\n",
    "import numpy as np\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "11eb08ba",
   "metadata": {},
   "outputs": [],
   "source": [
    "sys.path.append('../')\n",
    "from filter_clinvar_xml import filter_xml"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "6116614c",
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.set_option('display.max_colwidth', None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "624aa9f8",
   "metadata": {},
   "outputs": [],
   "source": [
    "PROJECT_ROOT = '/home/april/projects/opentargets'\n",
    "clinvar_path = os.path.join(PROJECT_ROOT, 'ClinVarFullRelease_00-latest.xml.gz')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "2b477981",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Only run this cell if you want to refresh clinvar data\n",
    "\n",
    "# clinvar_url = 'https://ftp.ncbi.nlm.nih.gov/pub/clinvar/xml/ClinVarFullRelease_00-latest.xml.gz'\n",
    "# request.urlretrieve(clinvar_url, clinvar_path)\n",
    "# request.urlcleanup()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d28b5fa0",
   "metadata": {},
   "source": [
    "## ClinVar documentation\n",
    "\n",
    "Submission guidelines can be found [here](https://www.ncbi.nlm.nih.gov/clinvar/docs/submit/#required), in particular note the requirement:\n",
    "\n",
    "> a valid description of the variant, one of:\n",
    ">    * an HGVS expression\n",
    ">    * chromosome coordinates and change\n",
    ">    * cytogenetic description\n",
    "\n",
    "Also found the [xsd](https://ftp.ncbi.nlm.nih.gov/pub/clinvar/clinvar_submission.xsd) for ClinVar XML submissions, which includes all possible measure types:\n",
    "\n",
    "```xml\n",
    "<xs:simpleType name=\"Measuretype\">\n",
    "    <xs:restriction base=\"xs:string\">\n",
    "        <xs:enumeration value=\"Gene\"/>\n",
    "        <xs:enumeration value=\"Variation\"/>\n",
    "        <xs:enumeration value=\"Insertion\"/>\n",
    "        <xs:enumeration value=\"Mobile element insertion\"/>\n",
    "        <xs:enumeration value=\"Novel sequence insertion\"/>\n",
    "        <xs:enumeration value=\"Microsatellite\"/>\n",
    "        <xs:enumeration value=\"Deletion\"/>\n",
    "        <xs:enumeration value=\"single nucleotide variant\"/>\n",
    "        <xs:enumeration value=\"Multiple nucleotide variation\"/>\n",
    "        <xs:enumeration value=\"Indel\"/>\n",
    "        <xs:enumeration value=\"Duplication\"/>\n",
    "        <xs:enumeration value=\"Tandem duplication\"/>\n",
    "        <xs:enumeration value=\"copy number loss\"/>\n",
    "        <xs:enumeration value=\"copy number gain\"/>\n",
    "        <xs:enumeration value=\"protein only\"/>\n",
    "        <xs:enumeration value=\"Inversion\"/>\n",
    "        <xs:enumeration value=\"Translocation\"/>\n",
    "        <xs:enumeration value=\"Interchromosomal breakpoint\"/>\n",
    "        <xs:enumeration value=\"Intrachromosomal breakpoint\"/>\n",
    "        <xs:enumeration value=\"Complex\"/>\n",
    "    </xs:restriction>\n",
    "</xs:simpleType>\n",
    "```\n",
    "Compare measure types we've actually found in the data, [here](https://github.com/EBIvariation/eva-opentargets/tree/master/data-exploration/clinvar-variant-types)."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "64b312a4",
   "metadata": {},
   "source": [
    "## Filter\n",
    "\n",
    "Filter full dataset to get just a manageable (hopefully representative) sample of records with measures representing complex events"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "a6e0f687",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Records written: 3040\n"
     ]
    }
   ],
   "source": [
    "complex_xml = os.path.join(PROJECT_ROOT, 'complex-events.xml.gz')\n",
    "\n",
    "# get just \"complex events\"\n",
    "# Q: what's complex? -- complex == no full coordinates\n",
    "def complex_measures(x):\n",
    "    if x.measure:\n",
    "        return (\n",
    "            # smattering of all non SNV variants\n",
    "            (x.measure.variant_type.lower() not in {'single nucleotide variant'} and np.random.random() < 0.01)\n",
    "            # be sure to get the rare ones\n",
    "            or (x.measure.variant_type.lower() in {'tandem duplication', 'fusion', 'complex', 'translocation', 'inversion'})\n",
    "        )\n",
    "    return False\n",
    "\n",
    "filter_xml(\n",
    "    input_xml=clinvar_path,\n",
    "    output_xml=complex_xml,\n",
    "    filter_fct=complex_measures,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "23153ed0",
   "metadata": {},
   "outputs": [],
   "source": [
    "dataset = ClinVarDataset(complex_xml)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b43e4d7e",
   "metadata": {},
   "source": [
    "## Dataframe\n",
    "\n",
    "Load measures into a dataframe where columns are the properties."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "21604d9c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_measures(dataset):\n",
    "    for r in dataset:\n",
    "        if r.measure:\n",
    "            yield r.measure"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "72976e1a",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "for m in get_measures(dataset):\n",
    "    break\n",
    "    \n",
    "dir(m)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "d4b6b900",
   "metadata": {},
   "outputs": [],
   "source": [
    "# just all the properties\n",
    "props = [\n",
    " 'all_names',\n",
    " 'clinvar_record',\n",
    "#  'explicit_insertion_length',\n",
    "#  'has_complete_coordinates',\n",
    " 'hgnc_ids',\n",
    " 'hgvs',\n",
    " 'is_repeat_expansion_variant',\n",
    "#  'measure_xml',\n",
    " 'microsatellite_category',\n",
    " 'nsv_id',\n",
    " 'preferred_gene_symbols',\n",
    "#  'preferred_name',\n",
    " 'preferred_or_other_name',\n",
    "#  'pubmed_refs',\n",
    " 'rs_id',\n",
    "#  'sequence_location_helper',\n",
    "#  'toplevel_refseq_hgvs',\n",
    " 'variant_type',\n",
    " 'chr',\n",
    " 'vcf_alt',\n",
    "#  'vcf_full_coords',\n",
    " 'vcf_pos',\n",
    " 'vcf_ref'\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "3f711b2a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# replaces empty list with None\n",
    "measures = [[getattr(v, p) if getattr(v, p) != [] else None for p in props] for v in get_measures(dataset)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "cb294f5d",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.DataFrame(measures, columns=props)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "7f02101b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "all_names                      2999\n",
       "clinvar_record                 3040\n",
       "hgnc_ids                       2610\n",
       "hgvs                           2392\n",
       "is_repeat_expansion_variant    3040\n",
       "microsatellite_category         208\n",
       "nsv_id                          272\n",
       "preferred_gene_symbols         2613\n",
       "preferred_or_other_name        2999\n",
       "rs_id                           996\n",
       "variant_type                   3040\n",
       "chr                            2033\n",
       "vcf_alt                        1830\n",
       "vcf_pos                        1830\n",
       "vcf_ref                        1830\n",
       "dtype: int64"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "d848a808",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Complex',\n",
       " 'Deletion',\n",
       " 'Duplication',\n",
       " 'Indel',\n",
       " 'Insertion',\n",
       " 'Inversion',\n",
       " 'Microsatellite',\n",
       " 'Tandem duplication',\n",
       " 'Translocation',\n",
       " 'Variation',\n",
       " 'copy number gain',\n",
       " 'copy number loss',\n",
       " 'fusion',\n",
       " 'protein only'}"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "set(df['variant_type'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "999d3b7d",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>all_names</th>\n",
       "      <th>clinvar_record</th>\n",
       "      <th>hgnc_ids</th>\n",
       "      <th>hgvs</th>\n",
       "      <th>is_repeat_expansion_variant</th>\n",
       "      <th>microsatellite_category</th>\n",
       "      <th>nsv_id</th>\n",
       "      <th>preferred_gene_symbols</th>\n",
       "      <th>preferred_or_other_name</th>\n",
       "      <th>rs_id</th>\n",
       "      <th>variant_type</th>\n",
       "      <th>chr</th>\n",
       "      <th>vcf_alt</th>\n",
       "      <th>vcf_pos</th>\n",
       "      <th>vcf_ref</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>[46;XX;t(1;5)(p32.1;q31.3)]</td>\n",
       "      <td>ClinVarRecord object with accession RCV000258575</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>46;XX;t(1;5)(p32.1;q31.3)</td>\n",
       "      <td>None</td>\n",
       "      <td>Translocation</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>[46;XY;t(1;15)(p22.3;q15)]</td>\n",
       "      <td>ClinVarRecord object with accession RCV000258594</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>46;XY;t(1;15)(p22.3;q15)</td>\n",
       "      <td>None</td>\n",
       "      <td>Translocation</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>[46;XX;t(2;11)(p13.1;p15.5)dn]</td>\n",
       "      <td>ClinVarRecord object with accession RCV000258610</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>46;XX;t(2;11)(p13.1;p15.5)dn</td>\n",
       "      <td>None</td>\n",
       "      <td>Translocation</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>[46;XY;t(5;8)(q15;q22.1)]</td>\n",
       "      <td>ClinVarRecord object with accession RCV000258613</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>46;XY;t(5;8)(q15;q22.1)</td>\n",
       "      <td>None</td>\n",
       "      <td>Translocation</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>[46;XY;t(5;15)(q13.1;q13)]</td>\n",
       "      <td>ClinVarRecord object with accession RCV000258628</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>46;XY;t(5;15)(q13.1;q13)</td>\n",
       "      <td>None</td>\n",
       "      <td>Translocation</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2732</th>\n",
       "      <td>[t(6;9)(q23.3;p22.3)]</td>\n",
       "      <td>ClinVarRecord object with accession RCV000585780</td>\n",
       "      <td>[HGNC:7545, HGNC:7785]</td>\n",
       "      <td>None</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[MYB, NFIB]</td>\n",
       "      <td>t(6;9)(q23.3;p22.3)</td>\n",
       "      <td>None</td>\n",
       "      <td>Translocation</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2766</th>\n",
       "      <td>[t(X;8)(p22.13;q12.1)]</td>\n",
       "      <td>ClinVarRecord object with accession RCV000767228</td>\n",
       "      <td>[HGNC:28509, HGNC:20626]</td>\n",
       "      <td>None</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[BEND2, CHD7]</td>\n",
       "      <td>t(X;8)(p22.13;q12.1)</td>\n",
       "      <td>None</td>\n",
       "      <td>Translocation</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2767</th>\n",
       "      <td>[t(8;9)(q24.22;q34.13)]</td>\n",
       "      <td>ClinVarRecord object with accession RCV000767229</td>\n",
       "      <td>[HGNC:12362, HGNC:26572]</td>\n",
       "      <td>None</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[TSC1, TMEM71]</td>\n",
       "      <td>t(8;9)(q24.22;q34.13)</td>\n",
       "      <td>None</td>\n",
       "      <td>Translocation</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2772</th>\n",
       "      <td>[t(5;22)(q31.2;q12.1)]</td>\n",
       "      <td>ClinVarRecord object with accession RCV000785874</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>t(5;22)(q31.2;q12.1)</td>\n",
       "      <td>None</td>\n",
       "      <td>Translocation</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2820</th>\n",
       "      <td>None</td>\n",
       "      <td>ClinVarRecord object with accession RCV001256208</td>\n",
       "      <td>[HGNC:13487, HGNC:3689]</td>\n",
       "      <td>None</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[VPS35, FGFR2]</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Translocation</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>259 rows Ã— 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                           all_names  \\\n",
       "1        [46;XX;t(1;5)(p32.1;q31.3)]   \n",
       "2         [46;XY;t(1;15)(p22.3;q15)]   \n",
       "3     [46;XX;t(2;11)(p13.1;p15.5)dn]   \n",
       "4          [46;XY;t(5;8)(q15;q22.1)]   \n",
       "5         [46;XY;t(5;15)(q13.1;q13)]   \n",
       "...                              ...   \n",
       "2732           [t(6;9)(q23.3;p22.3)]   \n",
       "2766          [t(X;8)(p22.13;q12.1)]   \n",
       "2767         [t(8;9)(q24.22;q34.13)]   \n",
       "2772          [t(5;22)(q31.2;q12.1)]   \n",
       "2820                            None   \n",
       "\n",
       "                                        clinvar_record  \\\n",
       "1     ClinVarRecord object with accession RCV000258575   \n",
       "2     ClinVarRecord object with accession RCV000258594   \n",
       "3     ClinVarRecord object with accession RCV000258610   \n",
       "4     ClinVarRecord object with accession RCV000258613   \n",
       "5     ClinVarRecord object with accession RCV000258628   \n",
       "...                                                ...   \n",
       "2732  ClinVarRecord object with accession RCV000585780   \n",
       "2766  ClinVarRecord object with accession RCV000767228   \n",
       "2767  ClinVarRecord object with accession RCV000767229   \n",
       "2772  ClinVarRecord object with accession RCV000785874   \n",
       "2820  ClinVarRecord object with accession RCV001256208   \n",
       "\n",
       "                      hgnc_ids  hgvs  is_repeat_expansion_variant  \\\n",
       "1                         None  None                        False   \n",
       "2                         None  None                        False   \n",
       "3                         None  None                        False   \n",
       "4                         None  None                        False   \n",
       "5                         None  None                        False   \n",
       "...                        ...   ...                          ...   \n",
       "2732    [HGNC:7545, HGNC:7785]  None                        False   \n",
       "2766  [HGNC:28509, HGNC:20626]  None                        False   \n",
       "2767  [HGNC:12362, HGNC:26572]  None                        False   \n",
       "2772                      None  None                        False   \n",
       "2820   [HGNC:13487, HGNC:3689]  None                        False   \n",
       "\n",
       "     microsatellite_category nsv_id preferred_gene_symbols  \\\n",
       "1                       None   None                   None   \n",
       "2                       None   None                   None   \n",
       "3                       None   None                   None   \n",
       "4                       None   None                   None   \n",
       "5                       None   None                   None   \n",
       "...                      ...    ...                    ...   \n",
       "2732                    None   None            [MYB, NFIB]   \n",
       "2766                    None   None          [BEND2, CHD7]   \n",
       "2767                    None   None         [TSC1, TMEM71]   \n",
       "2772                    None   None                   None   \n",
       "2820                    None   None         [VPS35, FGFR2]   \n",
       "\n",
       "           preferred_or_other_name rs_id   variant_type   chr vcf_alt vcf_pos  \\\n",
       "1        46;XX;t(1;5)(p32.1;q31.3)  None  Translocation  None    None    None   \n",
       "2         46;XY;t(1;15)(p22.3;q15)  None  Translocation  None    None    None   \n",
       "3     46;XX;t(2;11)(p13.1;p15.5)dn  None  Translocation  None    None    None   \n",
       "4          46;XY;t(5;8)(q15;q22.1)  None  Translocation  None    None    None   \n",
       "5         46;XY;t(5;15)(q13.1;q13)  None  Translocation  None    None    None   \n",
       "...                            ...   ...            ...   ...     ...     ...   \n",
       "2732           t(6;9)(q23.3;p22.3)  None  Translocation  None    None    None   \n",
       "2766          t(X;8)(p22.13;q12.1)  None  Translocation  None    None    None   \n",
       "2767         t(8;9)(q24.22;q34.13)  None  Translocation  None    None    None   \n",
       "2772          t(5;22)(q31.2;q12.1)  None  Translocation  None    None    None   \n",
       "2820                          None  None  Translocation  None    None    None   \n",
       "\n",
       "     vcf_ref  \n",
       "1       None  \n",
       "2       None  \n",
       "3       None  \n",
       "4       None  \n",
       "5       None  \n",
       "...      ...  \n",
       "2732    None  \n",
       "2766    None  \n",
       "2767    None  \n",
       "2772    None  \n",
       "2820    None  \n",
       "\n",
       "[259 rows x 15 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[df.variant_type == 'Translocation']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2b7e4b5c",
   "metadata": {},
   "source": [
    "## XML\n",
    "\n",
    "Probably not all elements from the raw xml are captured in the object, can we get find anything there?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "ab36f672",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "def get_measure_xml_for_rcv(dataset, rcv):\n",
    "    for r in dataset:\n",
    "        if r.accession == rcv:\n",
    "            return r.measure.measure_xml\n",
    "\n",
    "\n",
    "# pretty print xml\n",
    "def pprint(x):\n",
    "    print(ElementTree.tostring(x, encoding='unicode'))\n",
    "\n",
    "\n",
    "def print_measure_xml_for_rcv(dataset, rcv):\n",
    "    x = get_measure_xml_for_rcv(dataset, rcv)\n",
    "    pprint(x)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "d11f4647",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<Measure Type=\"Duplication\" ID=\"1039262\">\n",
      "        <Name>\n",
      "          <ElementValue Type=\"Preferred\">NC_000023.10:g.(?_31514885)_(32430050_?)dup</ElementValue>\n",
      "        </Name>\n",
      "        <AttributeSet>\n",
      "          <Attribute Accession=\"NC_000023\" Version=\"10\" Change=\"g.(?_31514885)_(32430050_?)dup\" Type=\"HGVS, genomic, top level, previous\" integerValue=\"37\">NC_000023.10:g.(?_31514885)_(32430050_?)dup</Attribute>\n",
      "        </AttributeSet>\n",
      "        <CytogeneticLocation>Xp21.1</CytogeneticLocation>\n",
      "        <SequenceLocation Assembly=\"GRCh37\" AssemblyAccessionVersion=\"GCF_000001405.25\" AssemblyStatus=\"previous\" Chr=\"X\" Accession=\"NC_000023.10\" innerStart=\"31514885\" innerStop=\"32430050\" display_start=\"31514885\" display_stop=\"32430050\" variantLength=\"915166\" />\n",
      "        <MeasureRelationship Type=\"within single gene\">\n",
      "          <Name>\n",
      "            <ElementValue Type=\"Preferred\">dystrophin</ElementValue>\n",
      "          </Name>\n",
      "          <Symbol>\n",
      "            <ElementValue Type=\"Preferred\">DMD</ElementValue>\n",
      "          </Symbol>\n",
      "          <AttributeSet>\n",
      "            <Attribute dateValue=\"2019-11-20\" Type=\"Haploinsufficiency\">Sufficient evidence for dosage pathogenicity</Attribute>\n",
      "            <Citation>\n",
      "              <URL>https://www.ncbi.nlm.nih.gov/projects/dbvar/ISCA/isca_gene.cgi?sym=DMD</URL>\n",
      "            </Citation>\n",
      "          </AttributeSet>\n",
      "          <AttributeSet>\n",
      "            <Attribute dateValue=\"2019-11-20\" Type=\"Triplosensitivity\">No evidence available</Attribute>\n",
      "            <Citation>\n",
      "              <URL>https://www.ncbi.nlm.nih.gov/projects/dbvar/ISCA/isca_gene.cgi?sym=DMD</URL>\n",
      "            </Citation>\n",
      "          </AttributeSet>\n",
      "          <SequenceLocation Assembly=\"GRCh38\" AssemblyAccessionVersion=\"GCF_000001405.38\" AssemblyStatus=\"current\" Chr=\"X\" Accession=\"NC_000023.11\" start=\"31119219\" stop=\"33339460\" display_start=\"31119219\" display_stop=\"33339460\" Strand=\"-\" />\n",
      "          <SequenceLocation Assembly=\"GRCh37\" AssemblyAccessionVersion=\"GCF_000001405.25\" AssemblyStatus=\"previous\" Chr=\"X\" Accession=\"NC_000023.10\" start=\"31137344\" stop=\"33357725\" display_start=\"31137344\" display_stop=\"33357725\" variantLength=\"2220382\" Strand=\"-\" />\n",
      "          <XRef ID=\"1756\" DB=\"Gene\" />\n",
      "          <XRef Type=\"MIM\" ID=\"300377\" DB=\"OMIM\" />\n",
      "          <XRef ID=\"HGNC:2928\" DB=\"HGNC\" />\n",
      "        </MeasureRelationship>\n",
      "      </Measure>\n",
      "      \n"
     ]
    }
   ],
   "source": [
    "xml = get_measure_xml_for_rcv(dataset, 'RCV001372309')\n",
    "pprint(xml)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "91016edd",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<Measure Type=\"Inversion\" ID=\"966181\">\n",
      "        <Name>\n",
      "          <ElementValue Type=\"Preferred\">NM_001098.3(ACO2):c.433-2_433-1inv</ElementValue>\n",
      "        </Name>\n",
      "        <CanonicalSPDI>NC_000022.11:41511873:AG:CT</CanonicalSPDI>\n",
      "        <AttributeSet>\n",
      "          <Attribute Accession=\"NM_001098\" Version=\"3\" Change=\"c.433-2_433-1inv\" Type=\"HGVS, coding, RefSeq\" MANESelect=\"true\">NM_001098.3:c.433-2_433-1inv</Attribute>\n",
      "        </AttributeSet>\n",
      "        <AttributeSet>\n",
      "          <Attribute Change=\"g.47750_47751delinsCT\" Accession=\"NG_032143\" Version=\"1\" Type=\"HGVS, genomic\">NG_032143.1:g.47750_47751delinsCT</Attribute>\n",
      "        </AttributeSet>\n",
      "        <AttributeSet>\n",
      "          <Attribute Accession=\"NG_032143\" Version=\"1\" Change=\"g.47750_47751delinsCT\" Type=\"HGVS, genomic, RefSeqGene\">NG_032143.1:g.47750_47751delinsCT</Attribute>\n",
      "        </AttributeSet>\n",
      "        <AttributeSet>\n",
      "          <Attribute Accession=\"NG_032143\" Version=\"1\" Change=\"g.47750_47751inv\" Type=\"HGVS, genomic, RefSeqGene\">NG_032143.1:g.47750_47751inv</Attribute>\n",
      "        </AttributeSet>\n",
      "        <AttributeSet>\n",
      "          <Attribute Accession=\"NC_000022\" Version=\"11\" Change=\"g.41511874_41511875inv\" Type=\"HGVS, genomic, top level\" integerValue=\"38\">NC_000022.11:g.41511874_41511875inv</Attribute>\n",
      "        </AttributeSet>\n",
      "        <AttributeSet>\n",
      "          <Attribute Accession=\"NC_000022\" Version=\"10\" Change=\"g.41907878_41907879inv\" Type=\"HGVS, genomic, top level, previous\" integerValue=\"37\">NC_000022.10:g.41907878_41907879inv</Attribute>\n",
      "        </AttributeSet>\n",
      "        <AttributeSet>\n",
      "          <Attribute Type=\"HGVS, incomplete\">c.433-2_443-1delinsCT</Attribute>\n",
      "        </AttributeSet>\n",
      "        <AttributeSet>\n",
      "          <Attribute Accession=\"NM_001098\" Version=\"2\" Change=\"c.433-2_433-1delinsCT\" Type=\"HGVS, previous\">NM_001098.2:c.433-2_433-1delinsCT</Attribute>\n",
      "        </AttributeSet>\n",
      "        <AttributeSet>\n",
      "          <Attribute Type=\"MolecularConsequence\">splice acceptor variant</Attribute>\n",
      "          <XRef ID=\"SO:0001574\" DB=\"Sequence Ontology\" />\n",
      "          <XRef ID=\"NM_001098.3:c.433-2_433-1inv\" DB=\"RefSeq\" />\n",
      "        </AttributeSet>\n",
      "        <CytogeneticLocation>22q13.2</CytogeneticLocation>\n",
      "        <SequenceLocation Assembly=\"GRCh38\" AssemblyAccessionVersion=\"GCF_000001405.38\" AssemblyStatus=\"current\" Chr=\"22\" Accession=\"NC_000022.11\" start=\"41511874\" stop=\"41511875\" display_start=\"41511874\" display_stop=\"41511875\" variantLength=\"2\" positionVCF=\"41511874\" referenceAlleleVCF=\"AG\" alternateAlleleVCF=\"CT\" />\n",
      "        <SequenceLocation Assembly=\"GRCh37\" AssemblyAccessionVersion=\"GCF_000001405.25\" AssemblyStatus=\"previous\" Chr=\"22\" Accession=\"NC_000022.10\" start=\"41907878\" stop=\"41907879\" display_start=\"41907878\" display_stop=\"41907879\" variantLength=\"2\" positionVCF=\"41907878\" referenceAlleleVCF=\"AG\" alternateAlleleVCF=\"CT\" />\n",
      "        <MeasureRelationship Type=\"within single gene\">\n",
      "          <Name>\n",
      "            <ElementValue Type=\"Preferred\">aconitase 2</ElementValue>\n",
      "          </Name>\n",
      "          <Symbol>\n",
      "            <ElementValue Type=\"Preferred\">ACO2</ElementValue>\n",
      "          </Symbol>\n",
      "          <SequenceLocation Assembly=\"GRCh38\" AssemblyAccessionVersion=\"GCF_000001405.38\" AssemblyStatus=\"current\" Chr=\"22\" Accession=\"NC_000022.11\" start=\"41468756\" stop=\"41528979\" display_start=\"41468756\" display_stop=\"41528979\" Strand=\"+\" />\n",
      "          <SequenceLocation Assembly=\"GRCh37\" AssemblyAccessionVersion=\"GCF_000001405.25\" AssemblyStatus=\"previous\" Chr=\"22\" Accession=\"NC_000022.10\" start=\"41865128\" stop=\"41924992\" display_start=\"41865128\" display_stop=\"41924992\" variantLength=\"59865\" Strand=\"+\" />\n",
      "          <XRef ID=\"50\" DB=\"Gene\" />\n",
      "          <XRef Type=\"MIM\" ID=\"100850\" DB=\"OMIM\" />\n",
      "          <XRef ID=\"HGNC:118\" DB=\"HGNC\" />\n",
      "        </MeasureRelationship>\n",
      "        <XRef Type=\"Allelic variant\" ID=\"100850.0010\" DB=\"OMIM\" />\n",
      "      </Measure>\n",
      "      \n"
     ]
    }
   ],
   "source": [
    "xml2 = get_measure_xml_for_rcv(dataset, 'RCV001255994')\n",
    "pprint(xml2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ecc15009",
   "metadata": {},
   "source": [
    "## Summary for ticket\n",
    "\n",
    "### ClinVar documentation\n",
    "\n",
    "Submission guidelines can be found [here](https://www.ncbi.nlm.nih.gov/clinvar/docs/submit/#required), in particular note the requirement:\n",
    "> * a valid description of the variant, one of:\n",
    ">    * an HGVS expression\n",
    ">    * chromosome coordinates and change\n",
    ">    * cytogenetic description\n",
    "\n",
    "### Notes from 4 August meeting\n",
    "\n",
    "Tentatively define \"complex events\" as variants without full `chr_pos_ref_alt` coordinates, as these are variants that aren't handled by the core VEP-based functional consequence pipeline.  Some of these *are* handled by the specialised repeat expansion pipeline, so we can focus on the rest.\n",
    "\n",
    "To really deal with complex events, besides identifiers we also need to get *functional consequences*, otherwise they get dropped by the pipeline.\n",
    "\n",
    "Next steps:\n",
    "1. Collect more representative examples of complex events in ClinVar\n",
    "1. Look at distribution of records with no functional consequences\n",
    "    * How many have HGVS-esque identifiers, sequence location with start/stop position, cytogenetic location?\n",
    "    * Of those with no full coordinates, how many can we parse? How many can the official HGVS parser parse?\n",
    "    * How are unparseable HGVS identifiers failing?  Is there some systematic way people are using HGVS-esque identifiers that aren't _technically_ HGVS-compliant but we can still parse reliably?\n",
    "1. What does VEP require to get consequences? Biologically we should only need start/stop coordinates.\n",
    "1. Can we report variant/phenotype associations without functional consequences, similar to ontology mappings?  Is this still valuable to OT?\n",
    "    * Can we report functional consequences with some uncertainty?  For example some HGVS identifiers encode uncertain location, e.g. [here](https://www.ncbi.nlm.nih.gov/clinvar/variation/1062574/).\n",
    "\n",
    "(1) and (2) I've started doing and will post in a subsequent comment before the next meeting."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "89e2f9c2",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}