{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "ccd3c577-253b-4ae3-98ee-ce7096bd0242",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "\n",
    "sys.path.append('..')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "f1efb678-3b9d-4192-93fa-12d1079a0745",
   "metadata": {},
   "outputs": [],
   "source": [
    "from filter_clinvar_xml import filter_xml, pprint, iterate_cvs_from_xml"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "34635850-c526-4f18-a431-f808a4692d34",
   "metadata": {},
   "outputs": [],
   "source": [
    "import gzip\n",
    "import os\n",
    "import pandas as pd\n",
    "from cmat.clinvar_xml_io import *\n",
    "from cmat.clinvar_xml_io.xml_parsing import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "3f81dd54-1ac0-48ad-b8f8-5c33d6418785",
   "metadata": {},
   "outputs": [],
   "source": [
    "work_dir = '/home/april/projects/opentargets/sept-investigation'\n",
    "july_clinvar_xml = os.path.join(work_dir, 'ClinVarRCVRelease_2024-07.xml.gz')\n",
    "missing_rcvs_tsv = os.path.join(work_dir, 'missing_rcvs.tsv.gz')\n",
    "missing_rcvs_output = os.path.join(work_dir, 'missing_rcvs_still_in_clinvar.xml.gz')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "ac3cd35d-7347-4fbe-beca-853c28b557e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "missing_rcvs = set()\n",
    "\n",
    "with gzip.open(missing_rcvs_tsv, 'rt') as fh:\n",
    "    # skip header\n",
    "    next(fh)\n",
    "    for line in fh:\n",
    "        if line:\n",
    "            missing_rcvs.add(line.split('\\t')[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "cfc9ed80-1006-4d4c-a8a7-b8a037fdcbb0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "359701"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(missing_rcvs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "d98accf9-fc62-4d59-aa71-a5e4dc761dee",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:filter_clinvar_xml:Records written: 106\n"
     ]
    }
   ],
   "source": [
    "filter_xml(\n",
    "    input_xml=july_clinvar_xml,\n",
    "    output_xml=missing_rcvs_output,\n",
    "    filter_fct = lambda r: r.accession in missing_rcvs,\n",
    "    max_num=len(missing_rcvs)\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "5ce075b4-1dc8-487a-9434-6bc0a64362e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "from cmat.clinvar_xml_io.clinical_classification import MultipleClinicalClassificationsError"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "816782d5-d48a-465e-ab37-696c25569406",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<ReferenceClinVarAssertion ID=\"3959282\" DateLastUpdated=\"2024-06-23\" DateCreated=\"2021-11-20\">\n",
      "    <ClinVarAccession Acc=\"RCV001779975\" Version=\"2\" Type=\"RCV\" DateUpdated=\"2024-06-23\" DateCreated=\"2021-11-20\" />\n",
      "    <RecordStatus>current</RecordStatus>\n",
      "    <Classifications>\n",
      "      <GermlineClassification>\n",
      "        <ReviewStatus>no assertion criteria provided</ReviewStatus>\n",
      "        <Description SubmissionCount=\"1\">Pathogenic</Description>\n",
      "      </GermlineClassification>\n",
      "    </Classifications>\n",
      "    <Assertion Type=\"variation to disease\" />\n",
      "    <ObservedIn>\n",
      "      <Sample>\n",
      "        <Origin>germline</Origin>\n",
      "        <Species TaxonomyId=\"9606\">human</Species>\n",
      "        <AffectedStatus>yes</AffectedStatus>\n",
      "      </Sample>\n",
      "      <Method>\n",
      "        <MethodType>research</MethodType>\n",
      "      </Method>\n",
      "      <ObservedData ID=\"158753121\">\n",
      "        <Attribute integerValue=\"1\" Type=\"VariantChromosomes\" />\n",
      "      </ObservedData>\n",
      "    </ObservedIn>\n",
      "    <MeasureSet Type=\"Variant\" ID=\"1321891\" Acc=\"VCV001321891\" Version=\"1\">\n",
      "      <Measure Type=\"Deletion\" ID=\"1312171\">\n",
      "        <Name>\n",
      "          <ElementValue Type=\"Preferred\">NC_000001.11:g.173888460_174138926del</ElementValue>\n",
      "        </Name>\n",
      "        <AttributeSet>\n",
      "          <Attribute Accession=\"NC_000001\" Version=\"11\" Change=\"g.173888460_174138926del\" Type=\"HGVS, genomic, top level\" integerValue=\"38\">NC_000001.11:g.173888460_174138926del</Attribute>\n",
      "        </AttributeSet>\n",
      "        <CytogeneticLocation>1q25.1</CytogeneticLocation>\n",
      "        <SequenceLocation Assembly=\"GRCh38\" AssemblyAccessionVersion=\"GCF_000001405.38\" AssemblyStatus=\"current\" Chr=\"1\" Accession=\"NC_000001.11\" start=\"173888460\" stop=\"174138926\" display_start=\"173888460\" display_stop=\"174138926\" variantLength=\"250467\" />\n",
      "        <MeasureRelationship Type=\"genes overlapped by variant\">\n",
      "          <Name>\n",
      "            <ElementValue Type=\"Preferred\">ATAC-STARR-seq lymphoblastoid active region 2119</ElementValue>\n",
      "          </Name>\n",
      "          <Symbol>\n",
      "            <ElementValue Type=\"Preferred\">LOC129931942</ElementValue>\n",
      "          </Symbol>\n",
      "          <SequenceLocation Assembly=\"GRCh38\" AssemblyAccessionVersion=\"GCF_000001405.38\" AssemblyStatus=\"current\" Chr=\"1\" Accession=\"NC_000001.11\" start=\"174020983\" stop=\"174021102\" display_start=\"174020983\" display_stop=\"174021102\" Strand=\"+\" />\n",
      "          <XRef ID=\"129931942\" DB=\"Gene\" />\n",
      "        </MeasureRelationship>\n",
      "        <MeasureRelationship Type=\"genes overlapped by variant\">\n",
      "          <Name>\n",
      "            <ElementValue Type=\"Preferred\">ATAC-STARR-seq lymphoblastoid silent region 1558</ElementValue>\n",
      "          </Name>\n",
      "          <Symbol>\n",
      "            <ElementValue Type=\"Preferred\">LOC129931943</ElementValue>\n",
      "          </Symbol>\n",
      "          <SequenceLocation Assembly=\"GRCh38\" AssemblyAccessionVersion=\"GCF_000001405.38\" AssemblyStatus=\"current\" Chr=\"1\" Accession=\"NC_000001.11\" start=\"174021913\" stop=\"174022232\" display_start=\"174021913\" display_stop=\"174022232\" Strand=\"+\" />\n",
      "          <XRef ID=\"129931943\" DB=\"Gene\" />\n",
      "        </MeasureRelationship>\n",
      "        <MeasureRelationship Type=\"genes overlapped by variant\">\n",
      "          <Name>\n",
      "            <ElementValue Type=\"Preferred\">ATAC-STARR-seq lymphoblastoid silent region 1559</ElementValue>\n",
      "          </Name>\n",
      "          <Symbol>\n",
      "            <ElementValue Type=\"Preferred\">LOC129931944</ElementValue>\n",
      "          </Symbol>\n",
      "          <SequenceLocation Assembly=\"GRCh38\" AssemblyAccessionVersion=\"GCF_000001405.38\" AssemblyStatus=\"current\" Chr=\"1\" Accession=\"NC_000001.11\" start=\"174022343\" stop=\"174022592\" display_start=\"174022343\" display_stop=\"174022592\" Strand=\"+\" />\n",
      "          <XRef ID=\"129931944\" DB=\"Gene\" />\n",
      "        </MeasureRelationship>\n",
      "        <MeasureRelationship Type=\"genes overlapped by variant\">\n",
      "          <Name>\n",
      "            <ElementValue Type=\"Preferred\">ATAC-STARR-seq lymphoblastoid silent region 1560</ElementValue>\n",
      "          </Name>\n",
      "          <Symbol>\n",
      "            <ElementValue Type=\"Preferred\">LOC129931945</ElementValue>\n",
      "          </Symbol>\n",
      "          <SequenceLocation Assembly=\"GRCh38\" AssemblyAccessionVersion=\"GCF_000001405.38\" AssemblyStatus=\"current\" Chr=\"1\" Accession=\"NC_000001.11\" start=\"174114956\" stop=\"174115015\" display_start=\"174114956\" display_stop=\"174115015\" Strand=\"+\" />\n",
      "          <XRef ID=\"129931945\" DB=\"Gene\" />\n",
      "        </MeasureRelationship>\n",
      "        <MeasureRelationship Type=\"genes overlapped by variant\">\n",
      "          <Name>\n",
      "            <ElementValue Type=\"Preferred\">BRD4-independent group 4 enhancer GRCh37_chr1:173863193-173864392</ElementValue>\n",
      "          </Name>\n",
      "          <Symbol>\n",
      "            <ElementValue Type=\"Preferred\">LOC126805923</ElementValue>\n",
      "          </Symbol>\n",
      "          <SequenceLocation Assembly=\"GRCh38\" AssemblyAccessionVersion=\"GCF_000001405.38\" AssemblyStatus=\"current\" Chr=\"1\" Accession=\"NC_000001.11\" start=\"173894055\" stop=\"173895254\" display_start=\"173894055\" display_stop=\"173895254\" Strand=\"+\" />\n",
      "          <XRef ID=\"126805923\" DB=\"Gene\" />\n",
      "        </MeasureRelationship>\n",
      "        <MeasureRelationship Type=\"genes overlapped by variant\">\n",
      "          <Name>\n",
      "            <ElementValue Type=\"Preferred\">MED14-independent group 3 enhancer GRCh37_chr1:173915901-173917100</ElementValue>\n",
      "          </Name>\n",
      "          <Symbol>\n",
      "            <ElementValue Type=\"Preferred\">LOC126805924</ElementValue>\n",
      "          </Symbol>\n",
      "          <SequenceLocation Assembly=\"GRCh38\" AssemblyAccessionVersion=\"GCF_000001405.38\" AssemblyStatus=\"current\" Chr=\"1\" Accession=\"NC_000001.11\" start=\"173946763\" stop=\"173947962\" display_start=\"173946763\" display_stop=\"173947962\" Strand=\"+\" />\n",
      "          <XRef ID=\"126805924\" DB=\"Gene\" />\n",
      "        </MeasureRelationship>\n",
      "        <MeasureRelationship Type=\"genes overlapped by variant\">\n",
      "          <Name>\n",
      "            <ElementValue Type=\"Preferred\">MPRA-validated peak465 silencer</ElementValue>\n",
      "          </Name>\n",
      "          <Symbol>\n",
      "            <ElementValue Type=\"Preferred\">LOC129388636</ElementValue>\n",
      "          </Symbol>\n",
      "          <SequenceLocation Assembly=\"GRCh38\" AssemblyAccessionVersion=\"GCF_000001405.38\" AssemblyStatus=\"current\" Chr=\"1\" Accession=\"NC_000001.11\" start=\"174002937\" stop=\"174003137\" display_start=\"174002937\" display_stop=\"174003137\" Strand=\"+\" />\n",
      "          <XRef ID=\"129388636\" DB=\"Gene\" />\n",
      "        </MeasureRelationship>\n",
      "        <MeasureRelationship Type=\"genes overlapped by variant\">\n",
      "          <Name>\n",
      "            <ElementValue Type=\"Preferred\">MPRA-validated peak466 silencer</ElementValue>\n",
      "          </Name>\n",
      "          <Symbol>\n",
      "            <ElementValue Type=\"Preferred\">LOC129388637</ElementValue>\n",
      "          </Symbol>\n",
      "          <SequenceLocation Assembly=\"GRCh38\" AssemblyAccessionVersion=\"GCF_000001405.38\" AssemblyStatus=\"current\" Chr=\"1\" Accession=\"NC_000001.11\" start=\"174014957\" stop=\"174015157\" display_start=\"174014957\" display_stop=\"174015157\" Strand=\"+\" />\n",
      "          <XRef ID=\"129388637\" DB=\"Gene\" />\n",
      "        </MeasureRelationship>\n",
      "        <MeasureRelationship Type=\"genes overlapped by variant\">\n",
      "          <Name>\n",
      "            <ElementValue Type=\"Preferred\">RABGAP1L divergent transcript</ElementValue>\n",
      "          </Name>\n",
      "          <Symbol>\n",
      "            <ElementValue Type=\"Preferred\">RABGAP1L-DT</ElementValue>\n",
      "          </Symbol>\n",
      "          <SequenceLocation Assembly=\"GRCh38\" AssemblyAccessionVersion=\"GCF_000001405.38\" AssemblyStatus=\"current\" Chr=\"1\" Accession=\"NC_000001.11\" start=\"174121638\" stop=\"174159287\" display_start=\"174121638\" display_stop=\"174159287\" Strand=\"-\" />\n",
      "          <XRef ID=\"102724601\" DB=\"Gene\" />\n",
      "          <XRef ID=\"HGNC:54296\" DB=\"HGNC\" />\n",
      "        </MeasureRelationship>\n",
      "        <MeasureRelationship Type=\"genes overlapped by variant\">\n",
      "          <Name>\n",
      "            <ElementValue Type=\"Preferred\">RC3H1 divergent transcript</ElementValue>\n",
      "          </Name>\n",
      "          <Symbol>\n",
      "            <ElementValue Type=\"Preferred\">RC3H1-DT</ElementValue>\n",
      "          </Symbol>\n",
      "          <SequenceLocation Assembly=\"GRCh38\" AssemblyAccessionVersion=\"GCF_000001405.38\" AssemblyStatus=\"current\" Chr=\"1\" Accession=\"NC_000001.11\" start=\"174022483\" stop=\"174023190\" display_start=\"174022483\" display_stop=\"174023190\" Strand=\"+\" />\n",
      "          <XRef ID=\"105371620\" DB=\"Gene\" />\n",
      "          <XRef ID=\"HGNC:54295\" DB=\"HGNC\" />\n",
      "        </MeasureRelationship>\n",
      "        <MeasureRelationship Type=\"genes overlapped by variant\">\n",
      "          <Name>\n",
      "            <ElementValue Type=\"Preferred\">Sharpr-MPRA regulatory region 797</ElementValue>\n",
      "          </Name>\n",
      "          <Symbol>\n",
      "            <ElementValue Type=\"Preferred\">LOC122149309</ElementValue>\n",
      "          </Symbol>\n",
      "          <SequenceLocation Assembly=\"GRCh38\" AssemblyAccessionVersion=\"GCF_000001405.38\" AssemblyStatus=\"current\" Chr=\"1\" Accession=\"NC_000001.11\" start=\"174131291\" stop=\"174131585\" display_start=\"174131291\" display_stop=\"174131585\" Strand=\"+\" />\n",
      "          <XRef ID=\"122149309\" DB=\"Gene\" />\n",
      "        </MeasureRelationship>\n",
      "        <MeasureRelationship Type=\"genes overlapped by variant\">\n",
      "          <Name>\n",
      "            <ElementValue Type=\"Preferred\">ring finger and CCCH-type domains 1</ElementValue>\n",
      "          </Name>\n",
      "          <Symbol>\n",
      "            <ElementValue Type=\"Preferred\">RC3H1</ElementValue>\n",
      "          </Symbol>\n",
      "          <SequenceLocation Assembly=\"GRCh38\" AssemblyAccessionVersion=\"GCF_000001405.38\" AssemblyStatus=\"current\" Chr=\"1\" Accession=\"NC_000001.11\" start=\"173931084\" stop=\"174022357\" display_start=\"173931084\" display_stop=\"174022357\" Strand=\"-\" />\n",
      "          <SequenceLocation Assembly=\"GRCh37\" AssemblyAccessionVersion=\"GCF_000001405.25\" AssemblyStatus=\"previous\" Chr=\"1\" Accession=\"NC_000001.10\" start=\"173900351\" stop=\"173962209\" display_start=\"173900351\" display_stop=\"173962209\" variantLength=\"61859\" Strand=\"-\" />\n",
      "          <XRef ID=\"149041\" DB=\"Gene\" />\n",
      "          <XRef Type=\"MIM\" ID=\"609424\" DB=\"OMIM\" />\n",
      "          <XRef ID=\"HGNC:29434\" DB=\"HGNC\" />\n",
      "        </MeasureRelationship>\n",
      "        <MeasureRelationship Type=\"genes overlapped by variant\">\n",
      "          <Name>\n",
      "            <ElementValue Type=\"Preferred\">serpin family C member 1</ElementValue>\n",
      "          </Name>\n",
      "          <Symbol>\n",
      "            <ElementValue Type=\"Preferred\">SERPINC1</ElementValue>\n",
      "          </Symbol>\n",
      "          <SequenceLocation Assembly=\"GRCh38\" AssemblyAccessionVersion=\"GCF_000001405.38\" AssemblyStatus=\"current\" Chr=\"1\" Accession=\"NC_000001.11\" start=\"173903800\" stop=\"173917327\" display_start=\"173903800\" display_stop=\"173917327\" Strand=\"-\" />\n",
      "          <SequenceLocation Assembly=\"GRCh37\" AssemblyAccessionVersion=\"GCF_000001405.25\" AssemblyStatus=\"previous\" Chr=\"1\" Accession=\"NC_000001.10\" start=\"173872941\" stop=\"173886515\" display_start=\"173872941\" display_stop=\"173886515\" variantLength=\"13575\" Strand=\"-\" />\n",
      "          <XRef ID=\"462\" DB=\"Gene\" />\n",
      "          <XRef Type=\"MIM\" ID=\"107300\" DB=\"OMIM\" />\n",
      "          <XRef ID=\"HGNC:775\" DB=\"HGNC\" />\n",
      "        </MeasureRelationship>\n",
      "        <MeasureRelationship Type=\"genes overlapped by variant\">\n",
      "          <Name>\n",
      "            <ElementValue Type=\"Preferred\">zinc finger and BTB domain containing 37</ElementValue>\n",
      "          </Name>\n",
      "          <Symbol>\n",
      "            <ElementValue Type=\"Preferred\">ZBTB37</ElementValue>\n",
      "          </Symbol>\n",
      "          <SequenceLocation Assembly=\"GRCh38\" AssemblyAccessionVersion=\"GCF_000001405.38\" AssemblyStatus=\"current\" Chr=\"1\" Accession=\"NC_000001.11\" start=\"173868082\" stop=\"173903547\" display_start=\"173868082\" display_stop=\"173903547\" Strand=\"+\" />\n",
      "          <SequenceLocation Assembly=\"GRCh37\" AssemblyAccessionVersion=\"GCF_000001405.25\" AssemblyStatus=\"previous\" Chr=\"1\" Accession=\"NC_000001.10\" start=\"173837492\" stop=\"173855773\" display_start=\"173837492\" display_stop=\"173855773\" variantLength=\"18282\" Strand=\"+\" />\n",
      "          <XRef ID=\"84614\" DB=\"Gene\" />\n",
      "          <XRef ID=\"HGNC:28365\" DB=\"HGNC\" />\n",
      "        </MeasureRelationship>\n",
      "      </Measure>\n",
      "      <Name>\n",
      "        <ElementValue Type=\"Preferred\">NC_000001.11:g.173888460_174138926del</ElementValue>\n",
      "      </Name>\n",
      "    </MeasureSet>\n",
      "    <TraitSet Type=\"Disease\" ID=\"6715\">\n",
      "      <Trait ID=\"6634\" Type=\"Disease\">\n",
      "        <Name>\n",
      "          <ElementValue Type=\"Preferred\">Hereditary antithrombin deficiency</ElementValue>\n",
      "          <XRef ID=\"MONDO:0013144\" DB=\"MONDO\" />\n",
      "        </Name>\n",
      "        <Name>\n",
      "          <ElementValue Type=\"Alternate\">Antithrombin III deficiency</ElementValue>\n",
      "          <XRef ID=\"Antithrombin+III+Deficiency/526\" DB=\"Genetic Alliance\" />\n",
      "          <XRef ID=\"HP:0001976\" DB=\"Human Phenotype Ontology\" />\n",
      "          <XRef ID=\"36351005\" DB=\"SNOMED CT\" />\n",
      "        </Name>\n",
      "        <Name>\n",
      "          <ElementValue Type=\"Alternate\">Thrombophilia due to antithrombin III deficiency</ElementValue>\n",
      "        </Name>\n",
      "        <Name>\n",
      "          <ElementValue Type=\"Alternate\">Reduced antithrombin III activity</ElementValue>\n",
      "          <XRef ID=\"HP:0001976\" DB=\"Human Phenotype Ontology\" />\n",
      "        </Name>\n",
      "        <Name>\n",
      "          <ElementValue Type=\"Alternate\">Antithrombin deficiency</ElementValue>\n",
      "          <XRef ID=\"UWMG_9235799\" DB=\"CSER _CC_NCGL, University of Washington\" />\n",
      "        </Name>\n",
      "        <Symbol>\n",
      "          <ElementValue Type=\"Alternate\">AT3D</ElementValue>\n",
      "          <XRef Type=\"MIM\" ID=\"613118\" DB=\"OMIM\" />\n",
      "        </Symbol>\n",
      "        <Symbol>\n",
      "          <ElementValue Type=\"Alternate\">THPH7</ElementValue>\n",
      "          <XRef Type=\"MIM\" ID=\"613118\" DB=\"OMIM\" />\n",
      "        </Symbol>\n",
      "        <AttributeSet>\n",
      "          <Attribute Type=\"GARD id\" integerValue=\"6148\" />\n",
      "          <XRef ID=\"6148\" DB=\"Office of Rare Diseases\" />\n",
      "        </AttributeSet>\n",
      "        <Citation Type=\"general\" Abbrev=\"ACOG, 2010\">\n",
      "          <ID Source=\"PubMed\">20027064</ID>\n",
      "        </Citation>\n",
      "        <Citation Type=\"general\" Abbrev=\"BSH, 2022\">\n",
      "          <ID Source=\"PubMed\">35645034</ID>\n",
      "        </Citation>\n",
      "        <XRef ID=\"MONDO:0013144\" DB=\"MONDO\" />\n",
      "        <XRef ID=\"C0272375\" DB=\"MedGen\" />\n",
      "        <XRef Type=\"MIM\" ID=\"613118\" DB=\"OMIM\" />\n",
      "        <XRef Type=\"primary\" ID=\"HP:0001976\" DB=\"Human Phenotype Ontology\" />\n",
      "      </Trait>\n",
      "    </TraitSet>\n",
      "  </ReferenceClinVarAssertion>\n",
      "  \n"
     ]
    }
   ],
   "source": [
    "# Check that these all have clinical significance \"no classifications from unflagged records\" or are dropped for multiple classifications\n",
    "for record in ClinVarDataset(missing_rcvs_output):\n",
    "    try:\n",
    "        if record.clinical_significance_list == record.valid_clinical_significances:\n",
    "            pprint(record.rcv)\n",
    "    except MultipleClinicalClassificationsError:\n",
    "        continue"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6630c6c7-191f-4af8-b103-96eae507d5bd",
   "metadata": {},
   "source": [
    "This one is dropped because it covers too many genes:\n",
    "```\n",
    "evidence_string_generation_2742194.err:WARNING:root:Skipping variant NC_000001.11:g.173888460_174138926del with 4 target genes\n",
    "```\n",
    "Note it's also missing in past 2 submissions, so not sure why it's in this list."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f1282d36-ced3-4320-8422-5fdb9877b605",
   "metadata": {},
   "source": [
    "Summary of the list of from Irene:\n",
    "* Out of **359,701** missing RCVs\n",
    "  * All but **106** were removed entirely from the ClinVar XML\n",
    "    * Of these 106, all but **1** have been modified so they were dropped with good reason (either a submission was flagged for insufficient evidence, or multiple classifications (somatic/oncogeneic) were added which are not yet supported)\n",
    "      * The 1 exception is dropped for being a large structural variant (too many target genes) and AFAICT was missing in March and June as well, so not sure how it ended up in the list\n",
    "\n",
    "From this, we can see that ClinVar deleted (at least) 359701 - 106 = **359,595** RCVs but had a net gain of **214,802** RCVs (see [metrics](https://docs.google.com/spreadsheets/d/1g_4tHNWP4VIikH7Jb0ui5aNr0PiFgvscZYOe69g191k/edit?usp=sharing)). So at least **574,397** RCVs were added.\n",
    "These counts are consistent with the assumption that:\n",
    "* the deletions include the **225,537** RCVs that previously produced evidence strings (\"done\" in the metrics spreadsheet)\n",
    "* the additions include the **440,301** new RCVs that are dropped for having invalid trait names (\"fatal\")\n",
    "\n",
    "So this can account for both the drop in number of evidence strings as well as the increase in invalid trait names. In particular, there's no evidence that any RCV switched from having a valid to an invalid trait name.\n",
    "\n",
    "This still leaves the question of why so many records are being added with non-specific trait names, but I think it confirms the pipeline is working as expected."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e70b0d87-772e-4a67-9939-c3de3f158a02",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
