{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "4fff2ba4",
   "metadata": {},
   "source": [
    "## Functional consequences\n",
    "\n",
    "* What functional consequences are already in ClinVar?\n",
    "* Are these directly from the submitters or annotated by ClinVar?\n",
    "* How often are they present?\n",
    "* How does the quality and coverage of the annotations compare with our annotations?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bebecf4f",
   "metadata": {},
   "source": [
    "### Background info\n",
    "\n",
    "From the [original paper](https://academic.oup.com/nar/article/42/D1/D980/1051029?login=true) on data augmentation - in particular functional/molecular consequences, genes and location info:\n",
    "> Some of the data ClinVar reports related to variation are added by NCBI. These data are reported only as part of the aggregate record (accession starting with RCV), and can include alternate HGVS expressions, allele frequencies from the 1000 Genomes project (7) or GO-ESP (8), identifiers from dbSNP or dbVar, molecular consequences (e.g. nonsense/missense/frameshift) and location data (splice site, untranslated regions, cytogenetic band, gene symbols and names). Values for molecular consequence, type of variation and location relative to a gene are standardized by reference to identifiers from the Sequence Ontology (9).\n",
    "\n",
    "In [full submission spreadsheet](https://www.ncbi.nlm.nih.gov/clinvar/docs/submit/#spreadsheet): don't see anything about molecular / functional consequences, which is consistent with the above stating that NCBI adds these.\n",
    "On the other hand there is space for optional gene symbol:\n",
    "> This must be the preferred symbol in NCBI's Gene database, which corresponds to HGNC's preferred symbol when one is available. Use only gene symbols that represent functional genes; do not include symbols for pseudogenes or regulatory regions. Separate multiple symbols with a semicolon.\n",
    "Gene symbol should only be provided to indicate the gene-disease relationship supporting the variant interpretation. Gene symbol is not expected for CNVs or cytogenetic variants except to make a statement that a specific gene within the variant has a relationship to the interpreted condition."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0c9a472a",
   "metadata": {},
   "source": [
    "### Coverage in ClinVar"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "1efaf7a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "\n",
    "sys.path.append('..')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "35e04bfb",
   "metadata": {},
   "outputs": [],
   "source": [
    "from collections import Counter\n",
    "import csv\n",
    "import multiprocessing\n",
    "from random import random\n",
    "import requests\n",
    "\n",
    "from filter_clinvar_xml import filter_xml, pprint, iterate_cvs_from_xml\n",
    "from eva_cttv_pipeline.clinvar_xml_io.clinvar_xml_io import *\n",
    "from eva_cttv_pipeline.trait_mapping.oxo import OntologyUri"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "11afd9ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "f12689c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# January 2023\n",
    "clinvar_xml = '/home/april/projects/opentargets/clinvar.xml.gz'\n",
    "clinvar = ClinVarDataset(clinvar_xml)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "d5bedc63",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_values(counter, keys):\n",
    "    return [counter.get(k, 0) for k in keys]\n",
    "\n",
    "\n",
    "def print_counter(counter, keys=None):\n",
    "    if not keys:\n",
    "        keys = counter.keys()\n",
    "    l = len(max(keys, key=lambda x: len(x)))\n",
    "    for k, v in zip(keys, get_values(counter, keys)):\n",
    "        print(f'{k: <{l}} {v}')\n",
    "        \n",
    "\n",
    "def plot_counter(counter, title, keys=None):\n",
    "    if not keys:\n",
    "        keys = counter.keys()\n",
    "    plt.figure(figsize=(15,10))\n",
    "    plt.title(title)\n",
    "    bars = plt.bar(keys, get_values(counter, keys))\n",
    "    plt.bar_label(bars, padding=3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "ce62f69e",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "# Note that\n",
    "#  - not all molecular consequence terms are SO\n",
    "#  - (possibly) not all SO terms are molecular consequences\n",
    "\n",
    "annotation_counts = Counter()\n",
    "for record in clinvar:\n",
    "    has_molec = False\n",
    "    has_func = False\n",
    "    has_gene = False\n",
    "    if record.measure:\n",
    "        for attr_elt in find_elements(record.measure.measure_xml, './AttributeSet'):\n",
    "            # These are usually Sequence Ontology\n",
    "            if find_elements(attr_elt, './Attribute[@Type=\"MolecularConsequence\"]'):\n",
    "                has_molec = True\n",
    "            # These are usually Variation Ontology (but sometimes SO)\n",
    "            if find_elements(attr_elt, './Attribute[@Type=\"FunctionalConsequence\"]'):\n",
    "                has_func = True\n",
    "        if record.measure.hgnc_ids:\n",
    "            has_gene = True\n",
    "    if has_molec:\n",
    "        annotation_counts['molecular_consequences'] += 1\n",
    "    if has_func:\n",
    "        annotation_counts['functional_consequences'] += 1\n",
    "    if has_gene:\n",
    "        annotation_counts['has_gene'] += 1\n",
    "    if not (has_molec or has_func or has_gene):\n",
    "        annotation_counts['none'] += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "d36ab885",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "molecular_consequences  2198584\n",
      "has_gene                2285631\n",
      "functional_consequences 16091\n",
      "none                    16652\n"
     ]
    }
   ],
   "source": [
    "print_counter(annotation_counts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "c1aec19c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# from metrics\n",
    "num_rcvs = 2302323  # verified same as count direct from XML\n",
    "missing_conseq = 20598"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "5068ecfe",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.9549415959446177"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# ClinVar's coverage of consequences\n",
    "annotation_counts['molecular_consequences'] / num_rcvs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "1f930b46",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.9910533839083395"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Our coverage of consequences\n",
    "(num_rcvs - missing_conseq) / num_rcvs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2cd249e7",
   "metadata": {},
   "source": [
    "#### Notes on our processing\n",
    "\n",
    "* we use HGNC id when we query biomart (repeat expansion pipeline) but not for structural or snp consequences\n",
    "* we definitely need to do processing here even if ClinVar coverage is 100% because our downstream usage requires Ensembl gene ids\n",
    "* see also [this (brief) discussion](https://github.com/EBIvariation/eva-opentargets/issues/189#issuecomment-786486333)\n",
    "\n",
    "\n",
    "For ontologies:\n",
    "* if CV's obsolete, ok\n",
    "* if not, ours need to be \"better\" - EFO, more specific/accurate, etc. (do this later)\n",
    "\n",
    "For consequences:\n",
    "* can also check if obsolete\n",
    "* if not, compare our annotations\n",
    "    * e.g. sample 1000, take molecular consequence from clinvar & functional consequences from ours & visually inspect differences in precision, etc."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3a79f21d",
   "metadata": {},
   "source": [
    "### SO Terms analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "000233ed",
   "metadata": {},
   "outputs": [],
   "source": [
    "term_counts = Counter()\n",
    "for record in clinvar:\n",
    "    annotated = False\n",
    "    if record.measure:\n",
    "        so_elts = find_elements(record.measure.measure_xml, './AttributeSet/XRef[@DB=\"Sequence Ontology\"]')\n",
    "        if so_elts:\n",
    "            so_terms = [elt.attrib['ID'] for elt in so_elts]\n",
    "            term_counts.update(so_terms)            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "f6c24c32",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Counter({'SO:0001583': 6607031,\n",
       "         'SO:0001623': 231681,\n",
       "         'SO:0001589': 1583489,\n",
       "         'SO:0001619': 420869,\n",
       "         'SO:0001587': 821551,\n",
       "         'SO:0001627': 1944452,\n",
       "         'SO:0001574': 75734,\n",
       "         'SO:0001575': 100717,\n",
       "         'SO:0001822': 60273,\n",
       "         'SO:0001819': 3028604,\n",
       "         'SO:0001821': 22691,\n",
       "         'SO:0001624': 333397,\n",
       "         'SO:0002317': 9,\n",
       "         'SO:0002054': 832,\n",
       "         'SO:1000117': 86,\n",
       "         'SO:1000064': 34,\n",
       "         'SO:0001582': 10610,\n",
       "         'SO:0001820': 74447,\n",
       "         'SO:0002073': 4002,\n",
       "         'SO:0001578': 4697,\n",
       "         'SO:0001536': 403,\n",
       "         'SO:1000071': 940,\n",
       "         'SO:0001986': 457,\n",
       "         'SO:0001555': 21,\n",
       "         'SO:0001542': 2,\n",
       "         'SO:0002153': 690,\n",
       "         'SO:0002152': 168,\n",
       "         'SO:0002220': 503,\n",
       "         'SO:1000054': 243,\n",
       "         'SO:0001559': 4,\n",
       "         'SO:0001987': 93,\n",
       "         'SO:1000072': 56,\n",
       "         'SO:0001786': 6,\n",
       "         'SO:0002218': 2038,\n",
       "         'SO:1000184': 4,\n",
       "         'SO:0002219': 4756,\n",
       "         'SO:0002053': 192,\n",
       "         'SO:1000070': 4,\n",
       "         'SO:0001541': 25,\n",
       "         'SO:0002052': 47,\n",
       "         'SO:0002316': 3,\n",
       "         'SO:0001561': 5,\n",
       "         'SO:0000683': 2,\n",
       "         'SO:0001565': 2})"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "term_counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "341a5fdf",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Our processing includes these as well as repeat expansion terms\n",
    "vep_terms = {'SO:0001893', 'SO:0001574', 'SO:0001575', 'SO:0001587', 'SO:0001589', 'SO:0001578', 'SO:0002012', 'SO:0001889', 'SO:0001821', 'SO:0001822', 'SO:0001583', 'SO:0001818', 'SO:0001630', 'SO:0001787', 'SO:0002170', 'SO:0002169', 'SO:0001626', 'SO:0002019', 'SO:0001567', 'SO:0001819', 'SO:0001580', 'SO:0001620', 'SO:0001623', 'SO:0001624', 'SO:0001792', 'SO:0001627', 'SO:0001621', 'SO:0001619', 'SO:0001631', 'SO:0001632', 'SO:0001895', 'SO:0001892', 'SO:0001782', 'SO:0001894', 'SO:0001891', 'SO:0001907', 'SO:0001566', 'SO:0001906', 'SO:0001628'}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "7643e158",
   "metadata": {},
   "outputs": [],
   "source": [
    "cv_terms = set(term_counts.keys())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "82aa9fb0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "39"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(vep_terms)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "c64f875a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "44"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(cv_terms)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "989e1c4a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'SO:0000683',\n",
       " 'SO:0001536',\n",
       " 'SO:0001541',\n",
       " 'SO:0001542',\n",
       " 'SO:0001555',\n",
       " 'SO:0001559',\n",
       " 'SO:0001561',\n",
       " 'SO:0001565',\n",
       " 'SO:0001582',\n",
       " 'SO:0001786',\n",
       " 'SO:0001820',\n",
       " 'SO:0001986',\n",
       " 'SO:0001987',\n",
       " 'SO:0002052',\n",
       " 'SO:0002053',\n",
       " 'SO:0002054',\n",
       " 'SO:0002073',\n",
       " 'SO:0002152',\n",
       " 'SO:0002153',\n",
       " 'SO:0002218',\n",
       " 'SO:0002219',\n",
       " 'SO:0002220',\n",
       " 'SO:0002316',\n",
       " 'SO:0002317',\n",
       " 'SO:1000054',\n",
       " 'SO:1000064',\n",
       " 'SO:1000070',\n",
       " 'SO:1000071',\n",
       " 'SO:1000072',\n",
       " 'SO:1000117',\n",
       " 'SO:1000184'}"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cv_terms - vep_terms"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "adc8c7a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "CURRENT = 'current'\n",
    "OBSOLETE = 'obsolete'\n",
    "\n",
    "\n",
    "def is_current(term):\n",
    "    url = f'https://www.ebi.ac.uk/ols/api/terms?id={term}'\n",
    "    response = requests.get(url)\n",
    "    response.raise_for_status()\n",
    "    try:\n",
    "        data = response.json()['_embedded']['terms'][0]\n",
    "        return (not data['is_obsolete'])\n",
    "    except Exception as e:\n",
    "        print('Not found', term)\n",
    "        return False    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "d2645b18",
   "metadata": {},
   "outputs": [],
   "source": [
    "# check obsolesence for clinvar (none of ours should be obsolete!)\n",
    "clinvar_obsolete = Counter()\n",
    "for term in term_counts:\n",
    "    if is_current(term):\n",
    "        clinvar_obsolete[CURRENT] += term_counts[term]\n",
    "    else:\n",
    "        clinvar_obsolete[OBSOLETE] += term_counts[term]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "84d262ee",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Counter({'current': 15334503, 'obsolete': 1367})"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinvar_obsolete"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3d17d933",
   "metadata": {},
   "source": [
    "Notes:\n",
    "\n",
    "- very few are obsolete, which is as we'd assume\n",
    "- assuming none from VEP are obsolete"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f8957a1f",
   "metadata": {},
   "source": [
    "### Annotations comparison"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "b43ac162",
   "metadata": {},
   "outputs": [],
   "source": [
    "from eva_cttv_pipeline.evidence_string_generation.clinvar_to_evidence_strings import get_consequence_types\n",
    "from eva_cttv_pipeline.evidence_string_generation.consequence_type import process_consequence_type_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "2137825a",
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_xml = '/home/april/projects/opentargets/sample.xml.gz'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "64e8515f",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:filter_clinvar_xml:Records written: 1144\n"
     ]
    }
   ],
   "source": [
    "# Sample ~1000 random records from current clinvar\n",
    "filter_xml(\n",
    "    input_xml=clinvar_xml,\n",
    "    output_xml=sample_xml,\n",
    "    filter_fct=lambda _: random() < (1000/2000000),\n",
    "    max_num=2000\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "66329105",
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_clinvar = ClinVarDataset(sample_xml)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "3d107d2f",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:eva_cttv_pipeline.evidence_string_generation:Loading mapping rs -> ENSG/SOterms\n",
      "INFO:eva_cttv_pipeline.evidence_string_generation:1577073 rs->ENSG/SOterms mappings loaded\n"
     ]
    }
   ],
   "source": [
    "variant_to_gene_mappings = process_consequence_type_file('/home/april/projects/opentargets/consequences_combined.tsv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "2f17f10c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_clinvar_consequences(measure):\n",
    "    # Only SO terms here\n",
    "    results = {}\n",
    "    for attr_elt in find_elements(record.measure.measure_xml, './AttributeSet'):\n",
    "        for conseq_elt in find_elements(attr_elt, './Attribute[@Type=\"MolecularConsequence\"]'):\n",
    "            conseq_text = conseq_elt.text\n",
    "            so_elt = find_mandatory_unique_element(attr_elt, './XRef[@DB=\"Sequence Ontology\"]')\n",
    "            so_term = so_elt.attrib['ID']\n",
    "            results[so_term] = conseq_text\n",
    "    return results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "f45f7806",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "RCV000050183\n",
      "ClinVar: {'SO:0001583': 'missense variant', 'SO:0001619': 'non-coding transcript variant'}\n",
      "Ours:    {'SO:0001583': 'missense variant'}\n",
      "=============\n",
      "RCV000161803\n",
      "ClinVar: {}\n",
      "Ours:    {'SO:0001627': 'intron variant'}\n",
      "=============\n",
      "RCV000250371\n",
      "ClinVar: {'SO:0001627': 'intron variant'}\n",
      "Ours:    {'SO:0002169': 'splice polypyrimidine tract variant'}\n",
      "=============\n",
      "RCV000303006\n",
      "ClinVar: {'SO:0001624': '3 prime UTR variant', 'SO:0001619': 'non-coding transcript variant'}\n",
      "Ours:    {'SO:0001624': '3 prime UTR variant'}\n",
      "=============\n",
      "RCV000309958\n",
      "ClinVar: {'SO:0001623': '5 prime UTR variant', 'SO:0001619': 'non-coding transcript variant'}\n",
      "Ours:    {'SO:0001623': '5 prime UTR variant'}\n",
      "=============\n",
      "RCV000310470\n",
      "ClinVar: {'SO:0001623': '5 prime UTR variant', 'SO:0001627': 'intron variant'}\n",
      "Ours:    {'SO:0001623': '5 prime UTR variant'}\n",
      "=============\n",
      "RCV000330816\n",
      "ClinVar: {'SO:0001624': '3 prime UTR variant', 'SO:0001619': 'non-coding transcript variant'}\n",
      "Ours:    {'SO:0001627': 'intron variant', 'SO:0001624': '3 prime UTR variant'}\n",
      "=============\n",
      "RCV000407652\n",
      "ClinVar: {'SO:0001627': 'intron variant'}\n",
      "Ours:    {'SO:0002169': 'splice polypyrimidine tract variant'}\n",
      "=============\n",
      "RCV000432795\n",
      "ClinVar: {'SO:0001623': '5 prime UTR variant', 'SO:0001619': 'non-coding transcript variant'}\n",
      "Ours:    {'SO:0001623': '5 prime UTR variant'}\n",
      "=============\n",
      "RCV000436693\n",
      "ClinVar: {'SO:0001583': 'missense variant', 'SO:0001619': 'non-coding transcript variant'}\n",
      "Ours:    {'SO:0001583': 'missense variant'}\n",
      "=============\n",
      "RCV000483774\n",
      "ClinVar: {'SO:0001589': 'frameshift variant', 'SO:0001619': 'non-coding transcript variant'}\n",
      "Ours:    {'SO:0001589': 'frameshift variant'}\n",
      "=============\n",
      "RCV000488877\n",
      "ClinVar: {'SO:0001819': 'synonymous variant'}\n",
      "Ours:    {'SO:0001630': 'splice region variant'}\n",
      "=============\n",
      "RCV000493340\n",
      "ClinVar: {'SO:0001589': 'frameshift variant', 'SO:0001627': 'intron variant'}\n",
      "Ours:    {'SO:0001589': 'frameshift variant'}\n",
      "=============\n",
      "RCV000499940\n",
      "ClinVar: {'SO:0001619': 'non-coding transcript variant', 'SO:0001819': 'synonymous variant'}\n",
      "Ours:    {'SO:0001819': 'synonymous variant'}\n",
      "=============\n",
      "RCV000754481\n",
      "ClinVar: {}\n",
      "Ours:    {'SO:0001627': 'intron variant'}\n",
      "=============\n",
      "RCV000788554\n",
      "ClinVar: {'SO:0001619': 'non-coding transcript variant', 'SO:0001587': 'nonsense'}\n",
      "Ours:    {'SO:0001587': 'stop gained'}\n",
      "=============\n",
      "RCV001042130\n",
      "ClinVar: {'SO:0001627': 'intron variant'}\n",
      "Ours:    {'SO:0002170': 'splice donor region variant'}\n",
      "=============\n",
      "RCV001071253\n",
      "ClinVar: {'SO:0001627': 'intron variant'}\n",
      "Ours:    {'SO:0002170': 'splice donor region variant'}\n",
      "=============\n",
      "RCV001169167\n",
      "ClinVar: {'SO:0001619': 'non-coding transcript variant', 'SO:0001819': 'synonymous variant'}\n",
      "Ours:    {'SO:0001819': 'synonymous variant'}\n",
      "=============\n",
      "RCV001199541\n",
      "ClinVar: {'SO:0001627': 'intron variant', 'SO:0001587': 'nonsense'}\n",
      "Ours:    {'SO:0001587': 'stop gained'}\n",
      "=============\n",
      "RCV001550454\n",
      "ClinVar: {'SO:0001627': 'intron variant'}\n",
      "Ours:    {'SO:0002169': 'splice polypyrimidine tract variant'}\n",
      "=============\n",
      "RCV001615732\n",
      "ClinVar: {'SO:0001627': 'intron variant', 'SO:0001583': 'missense variant'}\n",
      "Ours:    {'SO:0001583': 'missense variant'}\n",
      "=============\n",
      "RCV001619479\n",
      "ClinVar: {'SO:0001627': 'intron variant'}\n",
      "Ours:    {}\n",
      "=============\n",
      "RCV001690068\n",
      "ClinVar: {'SO:0001624': '3 prime UTR variant'}\n",
      "Ours:    {'SO:0001627': 'intron variant', 'SO:0001624': '3 prime UTR variant'}\n",
      "=============\n",
      "RCV001714942\n",
      "ClinVar: {}\n",
      "Ours:    {'SO:0001631': 'upstream gene variant'}\n",
      "=============\n",
      "RCV001764823\n",
      "ClinVar: {'SO:0001627': 'intron variant', 'SO:0001583': 'missense variant'}\n",
      "Ours:    {'SO:0001583': 'missense variant'}\n",
      "=============\n",
      "RCV001800141\n",
      "ClinVar: {'SO:0001820': 'inframe_indel'}\n",
      "Ours:    {'SO:0001818': 'protein altering variant'}\n",
      "=============\n",
      "RCV001879680\n",
      "ClinVar: {'SO:0001627': 'intron variant'}\n",
      "Ours:    {'SO:0002169': 'splice polypyrimidine tract variant'}\n",
      "=============\n",
      "RCV001926024\n",
      "ClinVar: {'SO:0001583': 'missense variant', 'SO:0001619': 'non-coding transcript variant'}\n",
      "Ours:    {'SO:0001583': 'missense variant'}\n",
      "=============\n",
      "RCV001948110\n",
      "ClinVar: {'SO:0001619': 'non-coding transcript variant', 'SO:0001819': 'synonymous variant'}\n",
      "Ours:    {'SO:0001819': 'synonymous variant'}\n",
      "=============\n",
      "RCV002024047\n",
      "ClinVar: {'SO:0001627': 'intron variant'}\n",
      "Ours:    {'SO:0001787': 'splice donor 5th base variant'}\n",
      "=============\n",
      "RCV002028365\n",
      "ClinVar: {'SO:0001627': 'intron variant', 'SO:0001583': 'missense variant'}\n",
      "Ours:    {'SO:0001583': 'missense variant'}\n",
      "=============\n",
      "RCV001382842\n",
      "ClinVar: {'SO:0001589': 'frameshift variant', 'SO:0001627': 'intron variant'}\n",
      "Ours:    {'SO:0001589': 'frameshift variant'}\n",
      "=============\n",
      "RCV000652035\n",
      "ClinVar: {'SO:0001819': 'synonymous variant'}\n",
      "Ours:    {'SO:0001630': 'splice region variant'}\n",
      "=============\n",
      "RCV000984402\n",
      "ClinVar: {'SO:0001623': '5 prime UTR variant', 'SO:0001627': 'intron variant', 'SO:0001575': 'splice donor variant'}\n",
      "Ours:    {'SO:0001575': 'splice donor variant'}\n",
      "=============\n",
      "RCV001216377\n",
      "ClinVar: {'SO:0001619': 'non-coding transcript variant'}\n",
      "Ours:    {'SO:0001631': 'upstream gene variant'}\n",
      "=============\n",
      "RCV001229007\n",
      "ClinVar: {'SO:0001627': 'intron variant'}\n",
      "Ours:    {'SO:0002170': 'splice donor region variant'}\n",
      "=============\n",
      "RCV001363858\n",
      "ClinVar: {'SO:0001589': 'frameshift variant', 'SO:0001619': 'non-coding transcript variant'}\n",
      "Ours:    {'SO:0001589': 'frameshift variant'}\n",
      "=============\n",
      "RCV000488749\n",
      "ClinVar: {'SO:0001627': 'intron variant'}\n",
      "Ours:    {'SO:0002170': 'splice donor region variant'}\n",
      "=============\n",
      "RCV002245198\n",
      "ClinVar: {'SO:0001619': 'non-coding transcript variant', 'SO:0001819': 'synonymous variant'}\n",
      "Ours:    {'SO:0001819': 'synonymous variant'}\n",
      "=============\n",
      "RCV001723339\n",
      "ClinVar: {'SO:0001627': 'intron variant', 'SO:0001587': 'nonsense'}\n",
      "Ours:    {'SO:0001623': '5 prime UTR variant'}\n",
      "=============\n",
      "RCV000503658\n",
      "ClinVar: {'SO:0001583': 'missense variant', 'SO:0001619': 'non-coding transcript variant'}\n",
      "Ours:    {'SO:0001583': 'missense variant'}\n",
      "=============\n",
      "RCV001733490\n",
      "ClinVar: {'SO:0001623': '5 prime UTR variant', 'SO:0001627': 'intron variant', 'SO:0001619': 'non-coding transcript variant'}\n",
      "Ours:    {'SO:0001623': '5 prime UTR variant'}\n",
      "=============\n",
      "RCV001819183\n",
      "ClinVar: {'SO:0001627': 'intron variant'}\n",
      "Ours:    {'SO:0002169': 'splice polypyrimidine tract variant'}\n",
      "=============\n",
      "RCV001820477\n",
      "ClinVar: {'SO:0001627': 'intron variant'}\n",
      "Ours:    {'SO:0002169': 'splice polypyrimidine tract variant'}\n",
      "=============\n",
      "RCV001940981\n",
      "ClinVar: {'SO:0001627': 'intron variant', 'SO:0001575': 'splice donor variant'}\n",
      "Ours:    {'SO:0001575': 'splice donor variant'}\n",
      "=============\n",
      "RCV000402245\n",
      "ClinVar: {'SO:0001624': '3 prime UTR variant', 'SO:0001583': 'missense variant', 'SO:0001619': 'non-coding transcript variant'}\n",
      "Ours:    {'SO:0001583': 'missense variant'}\n",
      "=============\n",
      "RCV001566291\n",
      "ClinVar: {'SO:0001583': 'missense variant', 'SO:0001819': 'synonymous variant'}\n",
      "Ours:    {'SO:0001583': 'missense variant'}\n",
      "=============\n",
      "RCV000361550\n",
      "ClinVar: {'SO:0001624': '3 prime UTR variant', 'SO:0001619': 'non-coding transcript variant'}\n",
      "Ours:    {'SO:0001624': '3 prime UTR variant'}\n",
      "=============\n",
      "RCV000118512\n",
      "ClinVar: {'SO:0001583': 'missense variant'}\n",
      "Ours:    {'SO:0001583': 'missense variant', 'SO:0001627': 'intron variant'}\n",
      "=============\n",
      "RCV000252673\n",
      "ClinVar: {'SO:0001627': 'intron variant'}\n",
      "Ours:    {'SO:0002169': 'splice polypyrimidine tract variant'}\n",
      "=============\n",
      "RCV000303716\n",
      "ClinVar: {'SO:0001819': 'synonymous variant'}\n",
      "Ours:    {'SO:0001587': 'stop gained'}\n",
      "=============\n",
      "RCV000352760\n",
      "ClinVar: {'SO:0001627': 'intron variant'}\n",
      "Ours:    {'SO:0001630': 'splice region variant'}\n",
      "=============\n",
      "RCV000418386\n",
      "ClinVar: {'SO:0001627': 'intron variant'}\n",
      "Ours:    {'SO:0001583': 'missense variant'}\n",
      "=============\n",
      "RCV000518533\n",
      "ClinVar: {'SO:0001619': 'non-coding transcript variant', 'SO:0001819': 'synonymous variant'}\n",
      "Ours:    {'SO:0001819': 'synonymous variant'}\n",
      "=============\n",
      "RCV000519934\n",
      "ClinVar: {'SO:0001583': 'missense variant', 'SO:0001619': 'non-coding transcript variant'}\n",
      "Ours:    {'SO:0001583': 'missense variant'}\n",
      "=============\n",
      "RCV000534941\n",
      "ClinVar: {'SO:0001819': 'synonymous variant'}\n",
      "Ours:    {'SO:0001627': 'intron variant', 'SO:0001819': 'synonymous variant'}\n",
      "=============\n",
      "RCV000638325\n",
      "ClinVar: {'SO:0001627': 'intron variant', 'SO:0001583': 'missense variant'}\n",
      "Ours:    {'SO:0001583': 'missense variant'}\n",
      "=============\n",
      "RCV000802336\n",
      "ClinVar: {'SO:0001583': 'missense variant', 'SO:0001619': 'non-coding transcript variant'}\n",
      "Ours:    {'SO:0001583': 'missense variant'}\n",
      "=============\n",
      "RCV000879693\n",
      "ClinVar: {'SO:0001583': 'missense variant', 'SO:0001619': 'non-coding transcript variant'}\n",
      "Ours:    {'SO:0001583': 'missense variant'}\n",
      "=============\n",
      "RCV000883325\n",
      "ClinVar: {'SO:0001623': '5 prime UTR variant', 'SO:0001619': 'non-coding transcript variant', 'SO:0001819': 'synonymous variant'}\n",
      "Ours:    {'SO:0001819': 'synonymous variant'}\n",
      "=============\n",
      "RCV000888220\n",
      "ClinVar: {'SO:0001627': 'intron variant'}\n",
      "Ours:    {'SO:0002169': 'splice polypyrimidine tract variant'}\n",
      "=============\n",
      "RCV000890748\n",
      "ClinVar: {'SO:0001627': 'intron variant'}\n",
      "Ours:    {'SO:0001630': 'splice region variant'}\n",
      "=============\n",
      "RCV000898591\n",
      "ClinVar: {'SO:0001627': 'intron variant'}\n",
      "Ours:    {'SO:0002169': 'splice polypyrimidine tract variant'}\n",
      "=============\n",
      "RCV000899461\n",
      "ClinVar: {'SO:0001627': 'intron variant'}\n",
      "Ours:    {'SO:0002169': 'splice polypyrimidine tract variant'}\n",
      "=============\n",
      "RCV000905164\n",
      "ClinVar: {'SO:0001627': 'intron variant'}\n",
      "Ours:    {'SO:0002169': 'splice polypyrimidine tract variant'}\n",
      "=============\n",
      "RCV000913168\n",
      "ClinVar: {'SO:0001623': '5 prime UTR variant'}\n",
      "Ours:    {'SO:0001583': 'missense variant'}\n",
      "=============\n",
      "RCV000913728\n",
      "ClinVar: {'SO:0001627': 'intron variant'}\n",
      "Ours:    {'SO:0001630': 'splice region variant'}\n",
      "=============\n",
      "RCV000921861\n",
      "ClinVar: {'SO:0001819': 'synonymous variant'}\n",
      "Ours:    {'SO:0001630': 'splice region variant'}\n",
      "=============\n",
      "RCV000943383\n",
      "ClinVar: {'SO:0001619': 'non-coding transcript variant', 'SO:0001819': 'synonymous variant'}\n",
      "Ours:    {'SO:0001819': 'synonymous variant'}\n",
      "=============\n",
      "RCV000947853\n",
      "ClinVar: {'SO:0001627': 'intron variant', 'SO:0001819': 'synonymous variant'}\n",
      "Ours:    {'SO:0001819': 'synonymous variant'}\n",
      "=============\n",
      "RCV000951409\n",
      "ClinVar: {'SO:0001619': 'non-coding transcript variant', 'SO:0001819': 'synonymous variant'}\n",
      "Ours:    {'SO:0001819': 'synonymous variant'}\n",
      "=============\n",
      "RCV000964314\n",
      "ClinVar: {'SO:0001619': 'non-coding transcript variant', 'SO:0001819': 'synonymous variant'}\n",
      "Ours:    {'SO:0001819': 'synonymous variant'}\n",
      "=============\n",
      "RCV000969492\n",
      "ClinVar: {'SO:0001627': 'intron variant', 'SO:0001583': 'missense variant'}\n",
      "Ours:    {'SO:0001583': 'missense variant'}\n",
      "=============\n",
      "RCV001060238\n",
      "ClinVar: {'SO:0001627': 'intron variant'}\n",
      "Ours:    {'SO:0002170': 'splice donor region variant'}\n",
      "=============\n",
      "RCV001114592\n",
      "ClinVar: {'SO:0001627': 'intron variant'}\n",
      "Ours:    {'SO:0001630': 'splice region variant'}\n",
      "=============\n",
      "RCV001223937\n",
      "ClinVar: {'SO:0001619': 'non-coding transcript variant', 'SO:0001587': 'nonsense'}\n",
      "Ours:    {'SO:0001587': 'stop gained'}\n",
      "=============\n",
      "RCV001274034\n",
      "ClinVar: {'SO:0001627': 'intron variant'}\n",
      "Ours:    {'SO:0002169': 'splice polypyrimidine tract variant'}\n",
      "=============\n",
      "RCV001309424\n",
      "ClinVar: {'SO:0001623': '5 prime UTR variant', 'SO:0001583': 'missense variant'}\n",
      "Ours:    {'SO:0001583': 'missense variant'}\n",
      "=============\n",
      "RCV001343855\n",
      "ClinVar: {'SO:0001627': 'intron variant', 'SO:0001583': 'missense variant'}\n",
      "Ours:    {'SO:0001583': 'missense variant'}\n",
      "=============\n",
      "RCV001349671\n",
      "ClinVar: {'SO:0001619': 'non-coding transcript variant', 'SO:0001819': 'synonymous variant'}\n",
      "Ours:    {'SO:0001819': 'synonymous variant'}\n",
      "=============\n",
      "RCV001351701\n",
      "ClinVar: {'SO:0001624': '3 prime UTR variant', 'SO:0001583': 'missense variant'}\n",
      "Ours:    {'SO:0001583': 'missense variant'}\n",
      "=============\n",
      "RCV001363238\n",
      "ClinVar: {'SO:0001583': 'missense variant', 'SO:0001619': 'non-coding transcript variant'}\n",
      "Ours:    {'SO:0001583': 'missense variant'}\n",
      "=============\n",
      "RCV001363464\n",
      "ClinVar: {'SO:0001627': 'intron variant', 'SO:0001583': 'missense variant'}\n",
      "Ours:    {'SO:0001583': 'missense variant'}\n",
      "=============\n",
      "RCV001364195\n",
      "ClinVar: {'SO:0001583': 'missense variant', 'SO:0001619': 'non-coding transcript variant'}\n",
      "Ours:    {'SO:0001583': 'missense variant'}\n",
      "=============\n",
      "RCV001391927\n",
      "ClinVar: {'SO:0001619': 'non-coding transcript variant', 'SO:0001819': 'synonymous variant'}\n",
      "Ours:    {'SO:0001819': 'synonymous variant'}\n",
      "=============\n",
      "RCV001409173\n",
      "ClinVar: {'SO:0001624': '3 prime UTR variant', 'SO:0001619': 'non-coding transcript variant', 'SO:0001819': 'synonymous variant'}\n",
      "Ours:    {'SO:0002169': 'splice polypyrimidine tract variant'}\n",
      "=============\n",
      "RCV001412507\n",
      "ClinVar: {'SO:0001627': 'intron variant'}\n",
      "Ours:    {'SO:0002169': 'splice polypyrimidine tract variant'}\n",
      "=============\n",
      "RCV001430160\n",
      "ClinVar: {'SO:0001619': 'non-coding transcript variant', 'SO:0001819': 'synonymous variant'}\n",
      "Ours:    {'SO:0001819': 'synonymous variant'}\n",
      "=============\n",
      "RCV001445294\n",
      "ClinVar: {'SO:0001623': '5 prime UTR variant', 'SO:0001583': 'missense variant'}\n",
      "Ours:    {'SO:0001583': 'missense variant'}\n",
      "=============\n",
      "RCV001451144\n",
      "ClinVar: {'SO:0001624': '3 prime UTR variant', 'SO:0001819': 'synonymous variant'}\n",
      "Ours:    {'SO:0001819': 'synonymous variant'}\n",
      "=============\n",
      "RCV001461814\n",
      "ClinVar: {'SO:0001623': '5 prime UTR variant', 'SO:0001819': 'synonymous variant'}\n",
      "Ours:    {'SO:0001819': 'synonymous variant'}\n",
      "=============\n",
      "RCV001480265\n",
      "ClinVar: {'SO:0001627': 'intron variant', 'SO:0001819': 'synonymous variant'}\n",
      "Ours:    {'SO:0001819': 'synonymous variant'}\n",
      "=============\n",
      "RCV001483763\n",
      "ClinVar: {'SO:0001619': 'non-coding transcript variant', 'SO:0001819': 'synonymous variant'}\n",
      "Ours:    {'SO:0001819': 'synonymous variant'}\n",
      "=============\n",
      "RCV001504410\n",
      "ClinVar: {'SO:0001619': 'non-coding transcript variant', 'SO:0001819': 'synonymous variant'}\n",
      "Ours:    {'SO:0001819': 'synonymous variant'}\n",
      "=============\n",
      "RCV001515984\n",
      "ClinVar: {'SO:0001627': 'intron variant', 'SO:0001619': 'non-coding transcript variant', 'SO:0001819': 'synonymous variant'}\n",
      "Ours:    {'SO:0001819': 'synonymous variant'}\n",
      "=============\n",
      "RCV001520285\n",
      "ClinVar: {'SO:0001623': '5 prime UTR variant', 'SO:0001627': 'intron variant'}\n",
      "Ours:    {'SO:0001623': '5 prime UTR variant'}\n",
      "=============\n",
      "RCV001520830\n",
      "ClinVar: {'SO:0001627': 'intron variant'}\n",
      "Ours:    {'SO:0002169': 'splice polypyrimidine tract variant'}\n",
      "=============\n",
      "RCV001729819\n",
      "ClinVar: {'SO:0001624': '3 prime UTR variant', 'SO:0001587': 'nonsense'}\n",
      "Ours:    {'SO:0001587': 'stop gained'}\n",
      "=============\n",
      "RCV001825964\n",
      "ClinVar: {'SO:0001619': 'non-coding transcript variant'}\n",
      "Ours:    {'SO:0001631': 'upstream gene variant'}\n",
      "=============\n",
      "RCV001828337\n",
      "ClinVar: {'SO:0001627': 'intron variant'}\n",
      "Ours:    {'SO:0002169': 'splice polypyrimidine tract variant'}\n",
      "=============\n",
      "RCV001838937\n",
      "ClinVar: {'SO:0001619': 'non-coding transcript variant', 'SO:0001819': 'synonymous variant'}\n",
      "Ours:    {'SO:0001819': 'synonymous variant'}\n",
      "=============\n",
      "RCV001844036\n",
      "ClinVar: {'SO:0001583': 'missense variant', 'SO:0001619': 'non-coding transcript variant'}\n",
      "Ours:    {'SO:0001583': 'missense variant'}\n",
      "=============\n",
      "RCV001906952\n",
      "ClinVar: {'SO:0001583': 'missense variant', 'SO:0001619': 'non-coding transcript variant'}\n",
      "Ours:    {'SO:0001583': 'missense variant'}\n",
      "=============\n",
      "RCV001979950\n",
      "ClinVar: {'SO:0001627': 'intron variant'}\n",
      "Ours:    {'SO:0002169': 'splice polypyrimidine tract variant'}\n",
      "=============\n",
      "RCV001982142\n",
      "ClinVar: {'SO:0001583': 'missense variant', 'SO:0001619': 'non-coding transcript variant'}\n",
      "Ours:    {'SO:0001583': 'missense variant'}\n",
      "=============\n",
      "RCV001983201\n",
      "ClinVar: {'SO:0001583': 'missense variant', 'SO:0001619': 'non-coding transcript variant'}\n",
      "Ours:    {'SO:0001583': 'missense variant'}\n",
      "=============\n",
      "RCV002005110\n",
      "ClinVar: {'SO:0001624': '3 prime UTR variant', 'SO:0001583': 'missense variant'}\n",
      "Ours:    {'SO:0001583': 'missense variant'}\n",
      "=============\n",
      "RCV002014105\n",
      "ClinVar: {'SO:0001623': '5 prime UTR variant', 'SO:0001583': 'missense variant', 'SO:0001619': 'non-coding transcript variant'}\n",
      "Ours:    {'SO:0001583': 'missense variant'}\n",
      "=============\n",
      "RCV002045285\n",
      "ClinVar: {'SO:0001627': 'intron variant', 'SO:0001583': 'missense variant'}\n",
      "Ours:    {'SO:0001583': 'missense variant'}\n",
      "=============\n",
      "RCV002069779\n",
      "ClinVar: {'SO:0001619': 'non-coding transcript variant', 'SO:0001819': 'synonymous variant'}\n",
      "Ours:    {'SO:0001567': 'stop retained variant'}\n",
      "=============\n",
      "RCV002077313\n",
      "ClinVar: {'SO:0001619': 'non-coding transcript variant', 'SO:0001819': 'synonymous variant'}\n",
      "Ours:    {'SO:0001819': 'synonymous variant'}\n",
      "=============\n",
      "RCV002080980\n",
      "ClinVar: {'SO:0001627': 'intron variant'}\n",
      "Ours:    {'SO:0002169': 'splice polypyrimidine tract variant'}\n",
      "=============\n",
      "RCV002085488\n",
      "ClinVar: {'SO:0001627': 'intron variant'}\n",
      "Ours:    {'SO:0002169': 'splice polypyrimidine tract variant'}\n",
      "=============\n",
      "RCV002090636\n",
      "ClinVar: {'SO:0001627': 'intron variant'}\n",
      "Ours:    {'SO:0002169': 'splice polypyrimidine tract variant'}\n",
      "=============\n",
      "RCV002093435\n",
      "ClinVar: {'SO:0001627': 'intron variant'}\n",
      "Ours:    {'SO:0002169': 'splice polypyrimidine tract variant'}\n",
      "=============\n",
      "RCV002103486\n",
      "ClinVar: {'SO:0001627': 'intron variant'}\n",
      "Ours:    {'SO:0001630': 'splice region variant'}\n",
      "=============\n",
      "RCV002107134\n",
      "ClinVar: {'SO:0001627': 'intron variant'}\n",
      "Ours:    {'SO:0002169': 'splice polypyrimidine tract variant'}\n",
      "=============\n",
      "RCV002132888\n",
      "ClinVar: {'SO:0001627': 'intron variant', 'SO:0001819': 'synonymous variant'}\n",
      "Ours:    {'SO:0001819': 'synonymous variant'}\n",
      "=============\n",
      "RCV002135376\n",
      "ClinVar: {'SO:0001627': 'intron variant'}\n",
      "Ours:    {'SO:0002169': 'splice polypyrimidine tract variant'}\n",
      "=============\n",
      "RCV002136210\n",
      "ClinVar: {'SO:0001627': 'intron variant'}\n",
      "Ours:    {'SO:0002169': 'splice polypyrimidine tract variant'}\n",
      "=============\n",
      "RCV002169082\n",
      "ClinVar: {'SO:0001627': 'intron variant'}\n",
      "Ours:    {'SO:0001630': 'splice region variant'}\n",
      "=============\n",
      "RCV002175155\n",
      "ClinVar: {'SO:0001627': 'intron variant'}\n",
      "Ours:    {'SO:0002169': 'splice polypyrimidine tract variant'}\n",
      "=============\n",
      "RCV002181753\n",
      "ClinVar: {'SO:0001619': 'non-coding transcript variant', 'SO:0001819': 'synonymous variant'}\n",
      "Ours:    {'SO:0001819': 'synonymous variant'}\n",
      "=============\n",
      "RCV002183005\n",
      "ClinVar: {'SO:0001627': 'intron variant'}\n",
      "Ours:    {'SO:0001624': '3 prime UTR variant'}\n",
      "=============\n",
      "RCV002184106\n",
      "ClinVar: {'SO:0001627': 'intron variant'}\n",
      "Ours:    {'SO:0001630': 'splice region variant'}\n",
      "=============\n",
      "RCV002185201\n",
      "ClinVar: {'SO:0001627': 'intron variant'}\n",
      "Ours:    {'SO:0002169': 'splice polypyrimidine tract variant'}\n",
      "=============\n",
      "RCV002190334\n",
      "ClinVar: {'SO:0001627': 'intron variant'}\n",
      "Ours:    {'SO:0002169': 'splice polypyrimidine tract variant'}\n",
      "=============\n",
      "RCV002198591\n",
      "ClinVar: {'SO:0001627': 'intron variant'}\n",
      "Ours:    {'SO:0002169': 'splice polypyrimidine tract variant'}\n",
      "=============\n",
      "RCV002202993\n",
      "ClinVar: {'SO:0001627': 'intron variant'}\n",
      "Ours:    {'SO:0002169': 'splice polypyrimidine tract variant'}\n",
      "=============\n",
      "RCV002204646\n",
      "ClinVar: {'SO:0001627': 'intron variant'}\n",
      "Ours:    {'SO:0002169': 'splice polypyrimidine tract variant'}\n",
      "=============\n",
      "RCV002205523\n",
      "ClinVar: {'SO:0001627': 'intron variant'}\n",
      "Ours:    {'SO:0002169': 'splice polypyrimidine tract variant'}\n",
      "=============\n",
      "RCV002207535\n",
      "ClinVar: {'SO:0001627': 'intron variant', 'SO:0001619': 'non-coding transcript variant', 'SO:0001819': 'synonymous variant'}\n",
      "Ours:    {'SO:0001819': 'synonymous variant'}\n",
      "=============\n",
      "RCV002211673\n",
      "ClinVar: {'SO:0001619': 'non-coding transcript variant', 'SO:0001819': 'synonymous variant'}\n",
      "Ours:    {'SO:0002169': 'splice polypyrimidine tract variant'}\n",
      "=============\n",
      "RCV002219657\n",
      "ClinVar: {'SO:0001627': 'intron variant'}\n",
      "Ours:    {'SO:0002169': 'splice polypyrimidine tract variant'}\n",
      "=============\n",
      "RCV002275462\n",
      "ClinVar: {'SO:0001583': 'missense variant', 'SO:0001619': 'non-coding transcript variant'}\n",
      "Ours:    {'SO:0001583': 'missense variant'}\n",
      "=============\n",
      "RCV002275760\n",
      "ClinVar: {'SO:0001624': '3 prime UTR variant'}\n",
      "Ours:    {'SO:0001623': '5 prime UTR variant', 'SO:0001624': '3 prime UTR variant'}\n",
      "=============\n",
      "RCV002077918\n",
      "ClinVar: {'SO:0001819': 'synonymous variant'}\n",
      "Ours:    {'SO:0001630': 'splice region variant'}\n",
      "=============\n",
      "RCV000133182\n",
      "ClinVar: {'SO:0001623': '5 prime UTR variant', 'SO:0001583': 'missense variant'}\n",
      "Ours:    {'SO:0001583': 'missense variant'}\n",
      "=============\n",
      "RCV000196481\n",
      "ClinVar: {'SO:0001627': 'intron variant', 'SO:0001583': 'missense variant'}\n",
      "Ours:    {'SO:0001583': 'missense variant'}\n",
      "=============\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "RCV000552047\n",
      "ClinVar: {'SO:0001627': 'intron variant'}\n",
      "Ours:    {'SO:0002169': 'splice polypyrimidine tract variant'}\n",
      "=============\n",
      "RCV000624821\n",
      "ClinVar: {'SO:0001583': 'missense variant', 'SO:0001619': 'non-coding transcript variant'}\n",
      "Ours:    {'SO:0001583': 'missense variant'}\n",
      "=============\n",
      "RCV000806155\n",
      "ClinVar: {'SO:0001627': 'intron variant'}\n",
      "Ours:    {'SO:0002170': 'splice donor region variant'}\n",
      "=============\n",
      "RCV001168107\n",
      "ClinVar: {'SO:0001624': '3 prime UTR variant', 'SO:0001619': 'non-coding transcript variant'}\n",
      "Ours:    {'SO:0001624': '3 prime UTR variant'}\n",
      "=============\n",
      "RCV001248688\n",
      "ClinVar: {'SO:0001583': 'missense variant', 'SO:0001619': 'non-coding transcript variant'}\n",
      "Ours:    {'SO:0001583': 'missense variant'}\n",
      "=============\n",
      "RCV001305520\n",
      "ClinVar: {'SO:0001627': 'intron variant', 'SO:0001583': 'missense variant'}\n",
      "Ours:    {'SO:0001583': 'missense variant'}\n",
      "=============\n",
      "RCV001493744\n",
      "ClinVar: {'SO:0001619': 'non-coding transcript variant', 'SO:0001819': 'synonymous variant'}\n",
      "Ours:    {'SO:0001819': 'synonymous variant'}\n",
      "=============\n",
      "RCV001712884\n",
      "ClinVar: {'SO:0001627': 'intron variant'}\n",
      "Ours:    {'SO:0002169': 'splice polypyrimidine tract variant'}\n",
      "=============\n",
      "RCV001920262\n",
      "ClinVar: {'SO:0001583': 'missense variant', 'SO:0001619': 'non-coding transcript variant'}\n",
      "Ours:    {'SO:0001583': 'missense variant'}\n",
      "=============\n",
      "RCV000015667\n",
      "ClinVar: {'SO:0001589': 'frameshift variant', 'SO:0001627': 'intron variant', 'SO:0001619': 'non-coding transcript variant'}\n",
      "Ours:    {'SO:0001589': 'frameshift variant'}\n",
      "=============\n",
      "RCV000521627\n",
      "ClinVar: {'SO:0001583': 'missense variant', 'SO:0001619': 'non-coding transcript variant'}\n",
      "Ours:    {'SO:0001583': 'missense variant'}\n",
      "=============\n",
      "RCV000032003\n",
      "ClinVar: {'SO:0001624': '3 prime UTR variant', 'SO:0001589': 'frameshift variant'}\n",
      "Ours:    {'SO:0001589': 'frameshift variant'}\n",
      "=============\n",
      "RCV000958554\n",
      "ClinVar: {'SO:0001583': 'missense variant', 'SO:0001619': 'non-coding transcript variant'}\n",
      "Ours:    {'SO:0001583': 'missense variant'}\n",
      "=============\n",
      "RCV001092935\n",
      "ClinVar: {'SO:0001819': 'synonymous variant'}\n",
      "Ours:    {'SO:0001583': 'missense variant'}\n",
      "=============\n",
      "RCV001174523\n",
      "ClinVar: {'SO:0001627': 'intron variant', 'SO:0001583': 'missense variant'}\n",
      "Ours:    {'SO:0001583': 'missense variant'}\n",
      "=============\n",
      "RCV000147334\n",
      "ClinVar: {'SO:0001589': 'frameshift variant', 'SO:0001627': 'intron variant'}\n",
      "Ours:    {'SO:0001589': 'frameshift variant'}\n",
      "=============\n",
      "RCV002196457\n",
      "ClinVar: {'SO:0001583': 'missense variant', 'SO:0001619': 'non-coding transcript variant'}\n",
      "Ours:    {'SO:0001583': 'missense variant'}\n",
      "=============\n",
      "RCV000426801\n",
      "ClinVar: {'SO:0001627': 'intron variant', 'SO:0001819': 'synonymous variant'}\n",
      "Ours:    {'SO:0001819': 'synonymous variant'}\n",
      "=============\n",
      "RCV000485119\n",
      "ClinVar: {'SO:0001624': '3 prime UTR variant'}\n",
      "Ours:    {'SO:0001624': '3 prime UTR variant', 'SO:0001627': 'intron variant'}\n",
      "=============\n",
      "RCV000508888\n",
      "ClinVar: {'SO:0001587': 'nonsense'}\n",
      "Ours:    {'SO:0001589': 'frameshift variant'}\n",
      "=============\n",
      "RCV000927626\n",
      "ClinVar: {'SO:0001627': 'intron variant', 'SO:0001819': 'synonymous variant'}\n",
      "Ours:    {'SO:0001819': 'synonymous variant'}\n",
      "=============\n",
      "RCV001209649\n",
      "ClinVar: {'SO:0001822': 'inframe_deletion', 'SO:0001820': 'inframe_indel'}\n",
      "Ours:    {'SO:0001822': 'inframe deletion'}\n",
      "=============\n",
      "RCV001306383\n",
      "ClinVar: {'SO:0001627': 'intron variant', 'SO:0001583': 'missense variant', 'SO:0001619': 'non-coding transcript variant'}\n",
      "Ours:    {'SO:0001583': 'missense variant'}\n",
      "=============\n",
      "RCV001462117\n",
      "ClinVar: {'SO:0001619': 'non-coding transcript variant', 'SO:0001819': 'synonymous variant'}\n",
      "Ours:    {'SO:0001819': 'synonymous variant'}\n",
      "=============\n",
      "RCV001665554\n",
      "ClinVar: {'SO:0001583': 'missense variant', 'SO:0001619': 'non-coding transcript variant'}\n",
      "Ours:    {'SO:0001583': 'missense variant'}\n",
      "=============\n",
      "RCV001968686\n",
      "ClinVar: {'SO:0001627': 'intron variant', 'SO:0001583': 'missense variant', 'SO:0001619': 'non-coding transcript variant'}\n",
      "Ours:    {'SO:0001583': 'missense variant'}\n",
      "=============\n",
      "RCV002067864\n",
      "ClinVar: {'SO:0001627': 'intron variant'}\n",
      "Ours:    {'SO:0002169': 'splice polypyrimidine tract variant'}\n",
      "=============\n",
      "RCV002304721\n",
      "ClinVar: {'SO:0001583': 'missense variant', 'SO:0001619': 'non-coding transcript variant'}\n",
      "Ours:    {'SO:0001583': 'missense variant'}\n",
      "=============\n",
      "RCV001068968\n",
      "ClinVar: {'SO:0001627': 'intron variant'}\n",
      "Ours:    {'SO:0002169': 'splice polypyrimidine tract variant'}\n",
      "=============\n",
      "RCV001340817\n",
      "ClinVar: {'SO:0001627': 'intron variant'}\n",
      "Ours:    {'SO:0002170': 'splice donor region variant'}\n",
      "=============\n",
      "RCV001466417\n",
      "ClinVar: {'SO:0001819': 'synonymous variant'}\n",
      "Ours:    {'SO:0001630': 'splice region variant'}\n",
      "=============\n",
      "RCV002305495\n",
      "ClinVar: {'SO:0001583': 'missense variant'}\n",
      "Ours:    {'SO:0001583': 'missense variant', 'SO:0001627': 'intron variant'}\n",
      "=============\n",
      "RCV000230471\n",
      "ClinVar: {'SO:0001619': 'non-coding transcript variant', 'SO:0001819': 'synonymous variant'}\n",
      "Ours:    {'SO:0001819': 'synonymous variant'}\n",
      "=============\n",
      "RCV000468603\n",
      "ClinVar: {'SO:0001623': '5 prime UTR variant', 'SO:0001583': 'missense variant'}\n",
      "Ours:    {'SO:0001583': 'missense variant'}\n",
      "=============\n",
      "RCV000565594\n",
      "ClinVar: {'SO:0001627': 'intron variant'}\n",
      "Ours:    {'SO:0002170': 'splice donor region variant'}\n",
      "=============\n",
      "RCV000756902\n",
      "ClinVar: {'SO:0001624': '3 prime UTR variant', 'SO:0001583': 'missense variant'}\n",
      "Ours:    {'SO:0001583': 'missense variant'}\n",
      "=============\n",
      "RCV000869102\n",
      "ClinVar: {'SO:0001627': 'intron variant', 'SO:0001619': 'non-coding transcript variant', 'SO:0001819': 'synonymous variant'}\n",
      "Ours:    {'SO:0001819': 'synonymous variant'}\n",
      "=============\n",
      "RCV001002556\n",
      "ClinVar: {'SO:0001627': 'intron variant', 'SO:0001583': 'missense variant'}\n",
      "Ours:    {'SO:0001583': 'missense variant'}\n",
      "=============\n",
      "RCV001012643\n",
      "ClinVar: {'SO:0001627': 'intron variant'}\n",
      "Ours:    {'SO:0002169': 'splice polypyrimidine tract variant'}\n",
      "=============\n",
      "RCV001016488\n",
      "ClinVar: {'SO:0001627': 'intron variant', 'SO:0001583': 'missense variant'}\n",
      "Ours:    {'SO:0001583': 'missense variant'}\n",
      "=============\n",
      "RCV001053344\n",
      "ClinVar: {'SO:0001583': 'missense variant', 'SO:0001619': 'non-coding transcript variant'}\n",
      "Ours:    {'SO:0001583': 'missense variant'}\n",
      "=============\n",
      "RCV001071271\n",
      "ClinVar: {'SO:0001623': '5 prime UTR variant', 'SO:0001583': 'missense variant'}\n",
      "Ours:    {'SO:0001583': 'missense variant'}\n",
      "=============\n",
      "RCV001182601\n",
      "ClinVar: {'SO:0001583': 'missense variant', 'SO:0001619': 'non-coding transcript variant'}\n",
      "Ours:    {'SO:0001583': 'missense variant'}\n",
      "=============\n",
      "RCV001223352\n",
      "ClinVar: {'SO:0001623': '5 prime UTR variant', 'SO:0001583': 'missense variant'}\n",
      "Ours:    {'SO:0001583': 'missense variant'}\n",
      "=============\n",
      "RCV001428215\n",
      "ClinVar: {'SO:0001619': 'non-coding transcript variant', 'SO:0001819': 'synonymous variant'}\n",
      "Ours:    {'SO:0001819': 'synonymous variant'}\n",
      "=============\n",
      "RCV001699013\n",
      "ClinVar: {'SO:0001627': 'intron variant', 'SO:0001574': 'splice acceptor variant'}\n",
      "Ours:    {'SO:0001574': 'splice acceptor variant'}\n",
      "=============\n",
      "RCV001808349\n",
      "ClinVar: {'SO:0001583': 'missense variant', 'SO:0001619': 'non-coding transcript variant'}\n",
      "Ours:    {'SO:0001583': 'missense variant'}\n",
      "=============\n",
      "RCV002256795\n",
      "ClinVar: {'SO:0001619': 'non-coding transcript variant', 'SO:0001819': 'synonymous variant'}\n",
      "Ours:    {'SO:0001819': 'synonymous variant'}\n",
      "=============\n",
      "RCV002298540\n",
      "ClinVar: {'SO:0001624': '3 prime UTR variant', 'SO:0001583': 'missense variant', 'SO:0001619': 'non-coding transcript variant'}\n",
      "Ours:    {'SO:0001583': 'missense variant'}\n",
      "=============\n",
      "RCV002322818\n",
      "ClinVar: {'SO:0001627': 'intron variant', 'SO:0001819': 'synonymous variant'}\n",
      "Ours:    {'SO:0001819': 'synonymous variant'}\n",
      "=============\n",
      "RCV002339365\n",
      "ClinVar: {'SO:0001619': 'non-coding transcript variant', 'SO:0001819': 'synonymous variant'}\n",
      "Ours:    {'SO:0001819': 'synonymous variant'}\n",
      "=============\n",
      "RCV002351335\n",
      "ClinVar: {'SO:0001619': 'non-coding transcript variant', 'SO:0001819': 'synonymous variant'}\n",
      "Ours:    {'SO:0001819': 'synonymous variant'}\n",
      "=============\n",
      "RCV002361421\n",
      "ClinVar: {'SO:0001619': 'non-coding transcript variant', 'SO:0001819': 'synonymous variant'}\n",
      "Ours:    {'SO:0001819': 'synonymous variant'}\n",
      "=============\n",
      "RCV002372890\n",
      "ClinVar: {'SO:0001623': '5 prime UTR variant', 'SO:0001819': 'synonymous variant'}\n",
      "Ours:    {'SO:0001819': 'synonymous variant'}\n",
      "=============\n",
      "RCV002385698\n",
      "ClinVar: {'SO:0001583': 'missense variant', 'SO:0001619': 'non-coding transcript variant'}\n",
      "Ours:    {'SO:0001583': 'missense variant'}\n",
      "=============\n",
      "RCV002406889\n",
      "ClinVar: {'SO:0001627': 'intron variant', 'SO:0001819': 'synonymous variant'}\n",
      "Ours:    {'SO:0001819': 'synonymous variant'}\n",
      "=============\n",
      "RCV002427449\n",
      "ClinVar: {'SO:0001624': '3 prime UTR variant', 'SO:0001583': 'missense variant'}\n",
      "Ours:    {'SO:0001583': 'missense variant'}\n",
      "=============\n",
      "RCV002442646\n",
      "ClinVar: {'SO:0001583': 'missense variant', 'SO:0001619': 'non-coding transcript variant'}\n",
      "Ours:    {'SO:0001583': 'missense variant'}\n",
      "=============\n",
      "RCV002443887\n",
      "ClinVar: {'SO:0001623': '5 prime UTR variant', 'SO:0001819': 'synonymous variant'}\n",
      "Ours:    {'SO:0001630': 'splice region variant'}\n",
      "=============\n",
      "RCV002455702\n",
      "ClinVar: {'SO:0001589': 'frameshift variant'}\n",
      "Ours:    {'SO:0001589': 'frameshift variant', 'SO:0001627': 'intron variant'}\n",
      "=============\n",
      "RCV002457927\n",
      "ClinVar: {'SO:0001583': 'missense variant', 'SO:0001619': 'non-coding transcript variant'}\n",
      "Ours:    {'SO:0001583': 'missense variant'}\n",
      "=============\n",
      "RCV000308521\n",
      "ClinVar: {'SO:0001624': '3 prime UTR variant', 'SO:0001627': 'intron variant'}\n",
      "Ours:    {'SO:0001624': '3 prime UTR variant'}\n",
      "=============\n",
      "RCV000468415\n",
      "ClinVar: {'SO:0001583': 'missense variant', 'SO:0001619': 'non-coding transcript variant'}\n",
      "Ours:    {'SO:0001583': 'missense variant'}\n",
      "=============\n",
      "RCV000583343\n",
      "ClinVar: {'SO:0001623': '5 prime UTR variant', 'SO:0001627': 'intron variant', 'SO:0001819': 'synonymous variant'}\n",
      "Ours:    {'SO:0001819': 'synonymous variant'}\n",
      "=============\n",
      "RCV000626130\n",
      "ClinVar: {'SO:0001627': 'intron variant', 'SO:0001583': 'missense variant', 'SO:0001619': 'non-coding transcript variant'}\n",
      "Ours:    {'SO:0001583': 'missense variant'}\n",
      "=============\n",
      "RCV001036740\n",
      "ClinVar: {'SO:0001819': 'synonymous variant'}\n",
      "Ours:    {'SO:0001630': 'splice region variant'}\n",
      "=============\n",
      "RCV001202625\n",
      "ClinVar: {'SO:0001821': 'inframe_insertion', 'SO:0001619': 'non-coding transcript variant'}\n",
      "Ours:    {'SO:0001821': 'inframe insertion'}\n",
      "=============\n",
      "RCV001262765\n",
      "ClinVar: {'SO:0001623': '5 prime UTR variant', 'SO:0001583': 'missense variant'}\n",
      "Ours:    {'SO:0001583': 'missense variant'}\n",
      "=============\n",
      "RCV001899446\n",
      "ClinVar: {'SO:0001589': 'frameshift variant', 'SO:0001627': 'intron variant'}\n",
      "Ours:    {'SO:0001589': 'frameshift variant'}\n",
      "=============\n",
      "RCV002003527\n",
      "ClinVar: {'SO:0001627': 'intron variant', 'SO:0001583': 'missense variant'}\n",
      "Ours:    {'SO:0001583': 'missense variant'}\n",
      "=============\n",
      "RCV000701724\n",
      "ClinVar: {'SO:0001627': 'intron variant', 'SO:0001583': 'missense variant'}\n",
      "Ours:    {'SO:0001583': 'missense variant'}\n",
      "=============\n",
      "RCV000203959\n",
      "ClinVar: {'SO:0001822': 'inframe_deletion'}\n",
      "Ours:    {'SO:0001822': 'inframe deletion', 'SO:0001627': 'intron variant'}\n",
      "=============\n",
      "RCV000561257\n",
      "ClinVar: {'SO:0001819': 'synonymous variant'}\n",
      "Ours:    {'SO:0001819': 'synonymous variant', 'SO:0001627': 'intron variant'}\n",
      "=============\n",
      "RCV000579859\n",
      "ClinVar: {'SO:0001623': '5 prime UTR variant', 'SO:0001819': 'synonymous variant'}\n",
      "Ours:    {'SO:0001630': 'splice region variant'}\n",
      "=============\n",
      "RCV000774599\n",
      "ClinVar: {'SO:0001583': 'missense variant'}\n",
      "Ours:    {'SO:0001583': 'missense variant', 'SO:0001627': 'intron variant'}\n",
      "=============\n",
      "RCV001194988\n",
      "ClinVar: {'SO:0001589': 'frameshift variant'}\n",
      "Ours:    {'SO:0001627': 'intron variant', 'SO:0001589': 'frameshift variant'}\n",
      "=============\n",
      "RCV000372565\n",
      "ClinVar: {'SO:0001627': 'intron variant'}\n",
      "Ours:    {'SO:0002169': 'splice polypyrimidine tract variant'}\n",
      "=============\n",
      "RCV001448640\n",
      "ClinVar: {'SO:0001619': 'non-coding transcript variant', 'SO:0001819': 'synonymous variant'}\n",
      "Ours:    {'SO:0001819': 'synonymous variant'}\n",
      "=============\n",
      "RCV002106047\n",
      "ClinVar: {'SO:0001627': 'intron variant'}\n",
      "Ours:    {'SO:0002169': 'splice polypyrimidine tract variant'}\n",
      "=============\n",
      "RCV002214603\n",
      "ClinVar: {'SO:0001627': 'intron variant', 'SO:0001819': 'synonymous variant'}\n",
      "Ours:    {'SO:0001819': 'synonymous variant'}\n",
      "=============\n",
      "RCV000178250\n",
      "ClinVar: {'SO:0001623': '5 prime UTR variant', 'SO:0001627': 'intron variant', 'SO:0001583': 'missense variant'}\n",
      "Ours:    {'SO:0001583': 'missense variant'}\n",
      "=============\n",
      "RCV000372255\n",
      "ClinVar: {'SO:0001583': 'missense variant', 'SO:0001619': 'non-coding transcript variant'}\n",
      "Ours:    {'SO:0001583': 'missense variant'}\n",
      "=============\n",
      "RCV000539500\n",
      "ClinVar: {'SO:0001627': 'intron variant'}\n",
      "Ours:    {'SO:0002170': 'splice donor region variant'}\n",
      "=============\n",
      "RCV000656173\n",
      "ClinVar: {'SO:0001583': 'missense variant'}\n",
      "Ours:    {'SO:0001627': 'intron variant', 'SO:0001583': 'missense variant'}\n",
      "=============\n",
      "RCV000687622\n",
      "ClinVar: {'SO:0001583': 'missense variant', 'SO:0001619': 'non-coding transcript variant'}\n",
      "Ours:    {'SO:0001583': 'missense variant'}\n",
      "=============\n",
      "RCV000733717\n",
      "ClinVar: {'SO:0001627': 'intron variant', 'SO:0001583': 'missense variant'}\n",
      "Ours:    {'SO:0001583': 'missense variant'}\n",
      "=============\n",
      "RCV001073404\n",
      "ClinVar: {'SO:0001627': 'intron variant', 'SO:0001583': 'missense variant', 'SO:0001619': 'non-coding transcript variant'}\n",
      "Ours:    {'SO:0001583': 'missense variant'}\n",
      "=============\n",
      "RCV001110977\n",
      "ClinVar: {'SO:0001583': 'missense variant', 'SO:0001619': 'non-coding transcript variant'}\n",
      "Ours:    {'SO:0001583': 'missense variant'}\n",
      "=============\n",
      "RCV001618253\n",
      "ClinVar: {'SO:0001627': 'intron variant'}\n",
      "Ours:    {'SO:0002169': 'splice polypyrimidine tract variant'}\n",
      "=============\n",
      "RCV002133422\n",
      "ClinVar: {'SO:0001627': 'intron variant'}\n",
      "Ours:    {'SO:0002169': 'splice polypyrimidine tract variant'}\n",
      "=============\n",
      "RCV002234177\n",
      "ClinVar: {'SO:0001627': 'intron variant', 'SO:0001819': 'synonymous variant'}\n",
      "Ours:    {'SO:0001819': 'synonymous variant'}\n",
      "=============\n",
      "RCV000038020\n",
      "ClinVar: {}\n",
      "Ours:    {'SO:0001589': 'frameshift variant'}\n",
      "=============\n",
      "RCV000039894\n",
      "ClinVar: {'SO:0001627': 'intron variant', 'SO:0001583': 'missense variant'}\n",
      "Ours:    {'SO:0001583': 'missense variant'}\n",
      "=============\n",
      "RCV001179341\n",
      "ClinVar: {'SO:0001819': 'synonymous variant'}\n",
      "Ours:    {'SO:0001630': 'splice region variant'}\n",
      "=============\n",
      "RCV001180807\n",
      "ClinVar: {'SO:0001624': '3 prime UTR variant', 'SO:0001589': 'frameshift variant', 'SO:0001627': 'intron variant'}\n",
      "Ours:    {'SO:0001624': '3 prime UTR variant'}\n",
      "=============\n",
      "RCV002071294\n",
      "ClinVar: {'SO:0001627': 'intron variant'}\n",
      "Ours:    {'SO:0002169': 'splice polypyrimidine tract variant'}\n",
      "=============\n",
      "RCV002371808\n",
      "ClinVar: {'SO:0001623': '5 prime UTR variant', 'SO:0001583': 'missense variant', 'SO:0001619': 'non-coding transcript variant'}\n",
      "Ours:    {'SO:0001583': 'missense variant'}\n",
      "=============\n",
      "RCV000086149\n",
      "ClinVar: {'SO:0001623': '5 prime UTR variant', 'SO:0001583': 'missense variant', 'SO:0001619': 'non-coding transcript variant'}\n",
      "Ours:    {'SO:0001583': 'missense variant'}\n",
      "=============\n",
      "RCV000408977\n",
      "ClinVar: {'SO:0001819': 'synonymous variant'}\n",
      "Ours:    {'SO:0001819': 'synonymous variant', 'SO:0001627': 'intron variant'}\n",
      "=============\n",
      "RCV000762839\n",
      "ClinVar: {'SO:0001619': 'non-coding transcript variant', 'SO:0001587': 'nonsense'}\n",
      "Ours:    {'SO:0001587': 'stop gained'}\n",
      "=============\n",
      "RCV001805033\n",
      "ClinVar: {'SO:0001589': 'frameshift variant', 'SO:0001627': 'intron variant'}\n",
      "Ours:    {'SO:0001589': 'frameshift variant'}\n",
      "=============\n",
      "RCV002470540\n",
      "ClinVar: {'SO:0001627': 'intron variant'}\n",
      "Ours:    {'SO:0002170': 'splice donor region variant'}\n",
      "=============\n",
      "RCV000014209\n",
      "ClinVar: {'SO:0001627': 'intron variant', 'SO:0001583': 'missense variant', 'SO:0001619': 'non-coding transcript variant'}\n",
      "Ours:    {'SO:0001583': 'missense variant'}\n",
      "=============\n",
      "RCV000034730\n",
      "ClinVar: {'SO:0001627': 'intron variant', 'SO:0001583': 'missense variant', 'SO:0001619': 'non-coding transcript variant'}\n",
      "Ours:    {'SO:0001583': 'missense variant'}\n",
      "=============\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "RCV000047838\n",
      "ClinVar: {'SO:0001589': 'frameshift variant', 'SO:0001627': 'intron variant', 'SO:0001619': 'non-coding transcript variant'}\n",
      "Ours:    {'SO:0001589': 'frameshift variant'}\n",
      "=============\n",
      "RCV000196528\n",
      "ClinVar: {}\n",
      "Ours:    {'SO:0001631': 'upstream gene variant'}\n",
      "=============\n",
      "RCV000345033\n",
      "ClinVar: {}\n",
      "Ours:    {'SO:0001623': '5 prime UTR variant'}\n",
      "=============\n",
      "RCV000355199\n",
      "ClinVar: {}\n",
      "Ours:    {'SO:0001631': 'upstream gene variant'}\n",
      "=============\n",
      "RCV000355884\n",
      "ClinVar: {'SO:0001627': 'intron variant', 'SO:0001583': 'missense variant'}\n",
      "Ours:    {'SO:0001583': 'missense variant'}\n",
      "=============\n",
      "RCV000359974\n",
      "ClinVar: {}\n",
      "Ours:    {'SO:0001623': '5 prime UTR variant'}\n",
      "=============\n",
      "RCV000369616\n",
      "ClinVar: {'SO:0001582': 'initiatior codon variant', 'SO:0001583': 'missense variant'}\n",
      "Ours:    {'SO:0002012': 'start lost'}\n",
      "=============\n",
      "RCV000389845\n",
      "ClinVar: {'SO:0001624': '3 prime UTR variant', 'SO:0001627': 'intron variant'}\n",
      "Ours:    {'SO:0001624': '3 prime UTR variant'}\n",
      "=============\n",
      "RCV000412601\n",
      "ClinVar: {'SO:0001589': 'frameshift variant', 'SO:0001627': 'intron variant', 'SO:0001619': 'non-coding transcript variant'}\n",
      "Ours:    {'SO:0001589': 'frameshift variant'}\n",
      "=============\n",
      "RCV000413040\n",
      "ClinVar: {'SO:0001627': 'intron variant'}\n",
      "Ours:    {'SO:0002170': 'splice donor region variant'}\n",
      "=============\n",
      "RCV000501898\n",
      "ClinVar: {'SO:0001589': 'frameshift variant', 'SO:0001627': 'intron variant', 'SO:0001619': 'non-coding transcript variant'}\n",
      "Ours:    {'SO:0001589': 'frameshift variant'}\n",
      "=============\n",
      "RCV000590751\n",
      "ClinVar: {'SO:0001589': 'frameshift variant', 'SO:0001627': 'intron variant', 'SO:0001619': 'non-coding transcript variant'}\n",
      "Ours:    {'SO:0001589': 'frameshift variant'}\n",
      "=============\n",
      "RCV000622343\n",
      "ClinVar: {'SO:0001587': 'nonsense'}\n",
      "Ours:    {'SO:0001589': 'frameshift variant'}\n",
      "=============\n",
      "RCV000675404\n",
      "ClinVar: {'SO:0001627': 'intron variant'}\n",
      "Ours:    {'SO:0002169': 'splice polypyrimidine tract variant'}\n",
      "=============\n",
      "RCV000722806\n",
      "ClinVar: {'SO:0001822': 'inframe_deletion', 'SO:0001627': 'intron variant', 'SO:0001619': 'non-coding transcript variant'}\n",
      "Ours:    {'SO:0001627': 'intron variant', 'SO:0001822': 'inframe deletion'}\n",
      "=============\n",
      "RCV000760065\n",
      "ClinVar: {'SO:0001583': 'missense variant', 'SO:0001619': 'non-coding transcript variant'}\n",
      "Ours:    {'SO:0001583': 'missense variant'}\n",
      "=============\n",
      "RCV000872844\n",
      "ClinVar: {'SO:0001627': 'intron variant'}\n",
      "Ours:    {'SO:0002169': 'splice polypyrimidine tract variant'}\n",
      "=============\n",
      "RCV000954594\n",
      "ClinVar: {'SO:0001583': 'missense variant', 'SO:0001619': 'non-coding transcript variant'}\n",
      "Ours:    {'SO:0001583': 'missense variant'}\n",
      "=============\n",
      "RCV001022804\n",
      "ClinVar: {'SO:0001589': 'frameshift variant', 'SO:0001619': 'non-coding transcript variant'}\n",
      "Ours:    {'SO:0001589': 'frameshift variant'}\n",
      "=============\n",
      "RCV001052016\n",
      "ClinVar: {'SO:0001627': 'intron variant', 'SO:0001619': 'non-coding transcript variant'}\n",
      "Ours:    {'SO:0002169': 'splice polypyrimidine tract variant', 'SO:0001620': 'mature miRNA variant'}\n",
      "=============\n",
      "RCV001084245\n",
      "ClinVar: {'SO:0001627': 'intron variant'}\n",
      "Ours:    {'SO:0002169': 'splice polypyrimidine tract variant'}\n",
      "=============\n",
      "RCV001110787\n",
      "ClinVar: {}\n",
      "Ours:    {'SO:0001623': '5 prime UTR variant'}\n",
      "=============\n",
      "RCV001129427\n",
      "ClinVar: {'SO:0001627': 'intron variant', 'SO:0001819': 'synonymous variant'}\n",
      "Ours:    {'SO:0001819': 'synonymous variant'}\n",
      "=============\n",
      "RCV001141180\n",
      "ClinVar: {}\n",
      "Ours:    {'SO:0001631': 'upstream gene variant'}\n",
      "=============\n",
      "RCV001169146\n",
      "ClinVar: {'SO:0001627': 'intron variant'}\n",
      "Ours:    {'SO:0001627': 'intron variant', 'SO:0001623': '5 prime UTR variant'}\n",
      "=============\n",
      "RCV001179205\n",
      "ClinVar: {'SO:0001627': 'intron variant'}\n",
      "Ours:    {'SO:0002169': 'splice polypyrimidine tract variant'}\n",
      "=============\n",
      "RCV001299844\n",
      "ClinVar: {'SO:0001624': '3 prime UTR variant', 'SO:0001627': 'intron variant', 'SO:0001583': 'missense variant'}\n",
      "Ours:    {'SO:0001583': 'missense variant'}\n",
      "=============\n",
      "RCV001406813\n",
      "ClinVar: {'SO:0001627': 'intron variant', 'SO:0001619': 'non-coding transcript variant', 'SO:0001819': 'synonymous variant'}\n",
      "Ours:    {'SO:0001819': 'synonymous variant'}\n",
      "=============\n",
      "RCV001653560\n",
      "ClinVar: {}\n",
      "Ours:    {'SO:0001631': 'upstream gene variant'}\n",
      "=============\n",
      "RCV001665549\n",
      "ClinVar: {'SO:0001627': 'intron variant'}\n",
      "Ours:    {'SO:0002169': 'splice polypyrimidine tract variant'}\n",
      "=============\n",
      "RCV001704260\n",
      "ClinVar: {'SO:0001627': 'intron variant'}\n",
      "Ours:    {'SO:0002170': 'splice donor region variant'}\n",
      "=============\n",
      "RCV001731920\n",
      "ClinVar: {'SO:0001627': 'intron variant', 'SO:0001583': 'missense variant'}\n",
      "Ours:    {'SO:0001627': 'intron variant'}\n",
      "=============\n",
      "RCV001766361\n",
      "ClinVar: {'SO:0001627': 'intron variant', 'SO:0001583': 'missense variant', 'SO:0001619': 'non-coding transcript variant'}\n",
      "Ours:    {'SO:0001583': 'missense variant'}\n",
      "=============\n",
      "RCV001798146\n",
      "ClinVar: {'SO:0001627': 'intron variant', 'SO:0001583': 'missense variant'}\n",
      "Ours:    {'SO:0001583': 'missense variant'}\n",
      "=============\n",
      "RCV001812266\n",
      "ClinVar: {'SO:0001623': '5 prime UTR variant', 'SO:0001589': 'frameshift variant'}\n",
      "Ours:    {'SO:0001589': 'frameshift variant'}\n",
      "=============\n",
      "RCV001873264\n",
      "ClinVar: {'SO:0001623': '5 prime UTR variant', 'SO:0001583': 'missense variant', 'SO:0001619': 'non-coding transcript variant'}\n",
      "Ours:    {'SO:0001623': '5 prime UTR variant', 'SO:0001583': 'missense variant'}\n",
      "=============\n",
      "RCV002235288\n",
      "ClinVar: {'SO:0001583': 'missense variant', 'SO:0001619': 'non-coding transcript variant'}\n",
      "Ours:    {'SO:0001583': 'missense variant'}\n",
      "=============\n",
      "RCV002243039\n",
      "ClinVar: {'SO:0001627': 'intron variant', 'SO:0001819': 'synonymous variant'}\n",
      "Ours:    {'SO:0001819': 'synonymous variant'}\n",
      "=============\n",
      "RCV002244728\n",
      "ClinVar: {'SO:0001623': '5 prime UTR variant', 'SO:0001587': 'nonsense'}\n",
      "Ours:    {'SO:0001587': 'stop gained'}\n",
      "=============\n",
      "RCV002318642\n",
      "ClinVar: {'SO:0001589': 'frameshift variant', 'SO:0001627': 'intron variant', 'SO:0001619': 'non-coding transcript variant'}\n",
      "Ours:    {'SO:0001589': 'frameshift variant'}\n",
      "=============\n",
      "RCV002325598\n",
      "ClinVar: {'SO:0001627': 'intron variant', 'SO:0001583': 'missense variant'}\n",
      "Ours:    {'SO:0001583': 'missense variant'}\n",
      "=============\n",
      "RCV002478378\n",
      "ClinVar: {'SO:0001627': 'intron variant'}\n",
      "Ours:    {'SO:0002170': 'splice donor region variant'}\n",
      "=============\n",
      "RCV002479867\n",
      "ClinVar: {'SO:0001627': 'intron variant'}\n",
      "Ours:    {'SO:0001583': 'missense variant'}\n",
      "=============\n",
      "RCV002481188\n",
      "ClinVar: {'SO:0001822': 'inframe_deletion', 'SO:0001619': 'non-coding transcript variant'}\n",
      "Ours:    {'SO:0001822': 'inframe deletion'}\n",
      "=============\n",
      "RCV002484348\n",
      "ClinVar: {'SO:0001627': 'intron variant', 'SO:0001583': 'missense variant'}\n",
      "Ours:    {'SO:0001583': 'missense variant'}\n",
      "=============\n",
      "RCV002489880\n",
      "ClinVar: {'SO:0001627': 'intron variant', 'SO:0001583': 'missense variant', 'SO:0001619': 'non-coding transcript variant'}\n",
      "Ours:    {'SO:0001583': 'missense variant'}\n",
      "=============\n",
      "RCV002502942\n",
      "ClinVar: {'SO:0001822': 'inframe_deletion', 'SO:0001627': 'intron variant'}\n",
      "Ours:    {'SO:0001822': 'inframe deletion'}\n",
      "=============\n"
     ]
    }
   ],
   "source": [
    "for record in sample_clinvar:\n",
    "    if record.measure:\n",
    "        clinvar_consequences = get_clinvar_consequences(record.measure)\n",
    "        our_consequences = get_consequence_types(record.measure, variant_to_gene_mappings)\n",
    "        our_consequences = {\n",
    "            c.so_term.accession.replace('_', ':'): c.so_term.so_name.replace('_', ' ')\n",
    "            for c in our_consequences\n",
    "        }\n",
    "\n",
    "        if clinvar_consequences.keys() != our_consequences.keys():\n",
    "            print(record.accession)\n",
    "            print('ClinVar:', clinvar_consequences)\n",
    "            print('Ours:   ', our_consequences)\n",
    "            print('=============')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2310d40f",
   "metadata": {},
   "source": [
    "#### More notes:\n",
    "\n",
    "* in ClinVar, consequences are associated with HGVS expressions as well, e.g. [RCV000050183](https://www.ncbi.nlm.nih.gov/clinvar/RCV000050183/)\n",
    "* our consequences are already filtered to include only most severe\n",
    "    * ClinVar may have some filtering as well but we don't know, and if anything it generally seems to include more (something like 1 consequence per HGVS expression (with a particular sequence identifier))\n",
    "* found at least one consequence from [Variation Ontology](http://www.variationontology.org/) rather than Sequence Ontology: [RCV000196528](https://www.ncbi.nlm.nih.gov/clinvar/RCV000196528/)\n",
    "* ClinVar uses nonsense instead of stop gained but apparently these are [synonymous](http://sequenceontology.org/browser/current_release/term/SO:0001587)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "993b90e3",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
